// @useoptic/standard-rulesets/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardRulesets = void 0;
const lintgpt_1 = require("./lintgpt");
__exportStar(require("./breaking-changes"), exports);
__exportStar(require("./naming-changes"), exports);
__exportStar(require("./examples"), exports);
__exportStar(require("./spectral"), exports);
__exportStar(require("./documentation"), exports);
const breaking_changes_1 = require("./breaking-changes");
const naming_changes_1 = require("./naming-changes");
const examples_1 = require("./examples");
const spectral_1 = require("./spectral");
const documentation_1 = require("./documentation");
exports.StandardRulesets = {
    'breaking-changes': breaking_changes_1.BreakingChangesRuleset,
    naming: naming_changes_1.NamingChangesRuleset,
    spectral: spectral_1.SpectralRulesets,
    examples: examples_1.ExamplesRuleset,
    documentation: documentation_1.DocumentationRuleset,
    lintgpt: lintgpt_1.LintGpt,
};
