// GendocuPublicApis/sdk/ts/gendocu/common/types_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/common/types.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

goog.exportSymbol('proto.gendocu.common.APISourceType', null, global);
goog.exportSymbol('proto.gendocu.common.ProgrammingLanguage', null, global);
goog.exportSymbol('proto.gendocu.common.SDKProvider', null, global);
/**
 * @enum {number}
 */
proto.gendocu.common.ProgrammingLanguage = {
  GO: 0,
  TYPESCRIPT: 1,
  PYTHON: 2,
  NODE_JS: 3,
  PHP: 4,
  JAVA: 5,
  CPP: 6,
  CSHARP: 7,
  DART: 8,
  KOTLIN: 9,
  OBJECTIVE_C: 10,
  RUBY: 11,
  GRPCURL: 12,
  REST_CURL: 13,
  JAVASCRIPT: 14,
  RUST: 15,
  UNKNOWN_PROGRAMMING_LANGUAGES: 255
};

/**
 * @enum {number}
 */
proto.gendocu.common.SDKProvider = {
  GENDOCU_GITEA: 0,
  PROTOFILE: 1,
  UNKNOWN_SDK_PROVIDER: 255
};

/**
 * @enum {number}
 */
proto.gendocu.common.APISourceType = {
  GITHUB: 0,
  PUBLIC: 1,
  UNKNOWN_API_SOURCE: 255
};

goog.object.extend(exports, proto.gendocu.common);
