// @useoptic/optic/build/commands/oas/specs/files/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpecFile = void 0;
Object.defineProperty(exports, "SpecFilesSourcemap", {
  enumerable: true,
  get: function () {
    return _openapiIo.JsonSchemaSourcemap;
  }
});
var _openapiIo = require("@useoptic/openapi-io");
var _reconcilers = require("./reconcilers");
var _stream = require("stream");
class SpecFile {
  static create(path) {
    return {
      path,
      contents: ''
    };
  }
  static async applyPatch(self, operations) {
    const result = await (0, _reconcilers.applyPatch)(self.path, self.contents, operations);
    if (!result.success) {
      throw new Error(`could not patch spec file: ${result.error}`);
    }
    return {
      path: self.path,
      contents: result.contents
    };
  }
  static isYaml(self) {
    return (0, _openapiIo.isYaml)(self.path);
  }
  static write(self, destination) {
    return _stream.Readable.from(self.contents).pipe(destination);
  }
}
exports.SpecFile = SpecFile;
const yamlCommentsPattern = /(^|[\s])#.*\n?/;
