// @useoptic/optic/build/commands/capture/patches/patchers/spec/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OperationDiffResultKind = void 0;
let OperationDiffResultKind = exports.OperationDiffResultKind = /*#__PURE__*/function (OperationDiffResultKind) {
  OperationDiffResultKind["UnmatchedPath"] = "UnmatchedPath";
  OperationDiffResultKind["UnmatchedPathParameter"] = "UnmatchedPathParameter";
  OperationDiffResultKind["UnmatchedMethod"] = "UnmatchedMethod";
  OperationDiffResultKind["MissingRequestBody"] = "MissingRequestBody";
  OperationDiffResultKind["UnmatchedRequestBody"] = "UnmatchedRequestBody";
  OperationDiffResultKind["MissingRequestParametersArray"] = "MissingRequestParametersArray";
  OperationDiffResultKind["UnmatchedRequestParameter"] = "UnmatchedRequestParameter";
  OperationDiffResultKind["MissingRequiredRequiredParameter"] = "MissingRequiredRequiredParameter";
  OperationDiffResultKind["MissingResponseBody"] = "MissingResponseBody";
  OperationDiffResultKind["UnmatchedResponseStatusCode"] = "UnmatchedResponseStatusCode";
  OperationDiffResultKind["UnmatchedResponseBody"] = "UnmatchdResponseBody";
  OperationDiffResultKind["MissingResponseHeadersObject"] = "MissingResponseHeadersObject";
  OperationDiffResultKind["UnmatchedResponseHeader"] = "UnmatchedResponseHeader";
  OperationDiffResultKind["MissingRequiredResponseHeader"] = "MissingRequiredResponseHeader";
  return OperationDiffResultKind;
}({});
