// @useoptic/optic/build/commands/oas/specs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FactVariants", {
  enumerable: true,
  get: function () {
    return _openapiUtilities.OpenApiKind;
  }
});
Object.defineProperty(exports, "OpenAPIV3", {
  enumerable: true,
  get: function () {
    return _openapiUtilities.OpenAPIV3;
  }
});
Object.defineProperty(exports, "SpecFile", {
  enumerable: true,
  get: function () {
    return _files.SpecFile;
  }
});
Object.defineProperty(exports, "SpecFileOperations", {
  enumerable: true,
  get: function () {
    return _files2.SpecFileOperations;
  }
});
Object.defineProperty(exports, "SpecFiles", {
  enumerable: true,
  get: function () {
    return _files2.SpecFiles;
  }
});
Object.defineProperty(exports, "SpecFilesAsync", {
  enumerable: true,
  get: function () {
    return _files2.SpecFilesAsync;
  }
});
Object.defineProperty(exports, "SpecFilesSourcemap", {
  enumerable: true,
  get: function () {
    return _files.SpecFilesSourcemap;
  }
});
Object.defineProperty(exports, "SpecTemplate", {
  enumerable: true,
  get: function () {
    return _templates.SpecTemplate;
  }
});
Object.defineProperty(exports, "isFactVariant", {
  enumerable: true,
  get: function () {
    return _openapiUtilities.isFactVariant;
  }
});
Object.defineProperty(exports, "readDeferencedSpec", {
  enumerable: true,
  get: function () {
    return _io.readDeferencedSpec;
  }
});
var _io = require("./io");
var _openapiUtilities = require("@useoptic/openapi-utilities");
var _files = require("./files");
var _files2 = require("./streams/files");
var _templates = require("./templates");
