// GendocuPublicApis/sdk/ts/gendocu/sdk_generator/v3/api_description_service_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/sdk_generator/v3/api_description_service.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

var gendocu_common_types_pb = require('../../../gendocu/common/types_pb.js');
goog.object.extend(proto, gendocu_common_types_pb);
var gendocu_sdk_generator_v3_sdk_generator_pb = require('../../../gendocu/sdk_generator/v3/sdk_generator_pb.js');
goog.object.extend(proto, gendocu_sdk_generator_v3_sdk_generator_pb);
var gendocu_sdk_generator_v3_authentication_model_pb = require('../../../gendocu/sdk_generator/v3/authentication_model_pb.js');
goog.object.extend(proto, gendocu_sdk_generator_v3_authentication_model_pb);
var google_api_field_behavior_pb = require('../../../google/api/field_behavior_pb.js');
goog.object.extend(proto, google_api_field_behavior_pb);
var google_protobuf_empty_pb = require('google-protobuf/google/protobuf/empty_pb.js');
goog.object.extend(proto, google_protobuf_empty_pb);
var google_protobuf_field_mask_pb = require('google-protobuf/google/protobuf/field_mask_pb.js');
goog.object.extend(proto, google_protobuf_field_mask_pb);
var google_protobuf_timestamp_pb = require('google-protobuf/google/protobuf/timestamp_pb.js');
goog.object.extend(proto, google_protobuf_timestamp_pb);
var gendocu_documentation_v1_styling_pb = require('../../../gendocu/documentation/v1/styling_pb.js');
goog.object.extend(proto, gendocu_documentation_v1_styling_pb);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.Build', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.BuildSelectorRequest', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.BuildSelectorRequest.FilterCase', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.BuildWithOrgAndName', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.CreateAPIDecode', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.EnterPasswordRequest', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.EnterPasswordResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.Environment', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.EnvironmentsRouting', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GenerateSDKResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GetDocumentationRequest', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.GetDocumentationResponse', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.Lines', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.ProvisioningInfo', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.SDKProvisioning', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.ServiceRouting', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.UpdateBuildRequest', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GetDocumentationRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GetDocumentationRequest.displayName = 'proto.gendocu.sdk_generator.v3.GetDocumentationRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GetDocumentationResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GetDocumentationResponse.displayName = 'proto.gendocu.sdk_generator.v3.GetDocumentationResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.EnterPasswordRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.EnterPasswordRequest.displayName = 'proto.gendocu.sdk_generator.v3.EnterPasswordRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.EnterPasswordResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.EnterPasswordResponse.displayName = 'proto.gendocu.sdk_generator.v3.EnterPasswordResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.displayName = 'proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.displayName = 'proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.Lines = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.Lines.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.Lines, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.Lines.displayName = 'proto.gendocu.sdk_generator.v3.Lines';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.displayName = 'proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.displayName = 'proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.CreateAPIDecode.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.CreateAPIDecode, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.CreateAPIDecode.displayName = 'proto.gendocu.sdk_generator.v3.CreateAPIDecode';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.EnvironmentsRouting, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.EnvironmentsRouting.displayName = 'proto.gendocu.sdk_generator.v3.EnvironmentsRouting';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.ServiceRouting = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.ServiceRouting.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.ServiceRouting, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.ServiceRouting.displayName = 'proto.gendocu.sdk_generator.v3.ServiceRouting';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.displayName = 'proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.GenerateSDKResponse.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.GenerateSDKResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.GenerateSDKResponse.displayName = 'proto.gendocu.sdk_generator.v3.GenerateSDKResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_);
};
goog.inherits(proto.gendocu.sdk_generator.v3.BuildSelectorRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.BuildSelectorRequest.displayName = 'proto.gendocu.sdk_generator.v3.BuildSelectorRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.BuildWithOrgAndName, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.displayName = 'proto.gendocu.sdk_generator.v3.BuildWithOrgAndName';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.Build = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.Build.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.Build, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.Build.displayName = 'proto.gendocu.sdk_generator.v3.Build';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.Environment = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.Environment, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.Environment.displayName = 'proto.gendocu.sdk_generator.v3.Environment';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.SDKProvisioning, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.SDKProvisioning.displayName = 'proto.gendocu.sdk_generator.v3.SDKProvisioning';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.ProvisioningInfo, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.ProvisioningInfo.displayName = 'proto.gendocu.sdk_generator.v3.ProvisioningInfo';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.UpdateBuildRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.UpdateBuildRequest.displayName = 'proto.gendocu.sdk_generator.v3.UpdateBuildRequest';
}



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GetDocumentationRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    selector: (f = msg.getSelector()) && proto.gendocu.sdk_generator.v3.BuildSelectorRequest.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GetDocumentationRequest;
  return proto.gendocu.sdk_generator.v3.GetDocumentationRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.BuildSelectorRequest;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.BuildSelectorRequest.deserializeBinaryFromReader);
      msg.setSelector(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GetDocumentationRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getSelector();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.BuildSelectorRequest.serializeBinaryToWriter
    );
  }
};


/**
 * optional BuildSelectorRequest selector = 1;
 * @return {?proto.gendocu.sdk_generator.v3.BuildSelectorRequest}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.getSelector = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.BuildSelectorRequest} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.BuildSelectorRequest, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.BuildSelectorRequest|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest} returns this
*/
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.setSelector = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.clearSelector = function() {
  return this.setSelector(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationRequest.prototype.hasSelector = function() {
  return jspb.Message.getField(this, 1) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GetDocumentationResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    build: (f = msg.getBuild()) && proto.gendocu.sdk_generator.v3.Build.toObject(includeInstance, f),
    isPrivate: jspb.Message.getBooleanFieldWithDefault(msg, 2, false),
    canEditDocumentation: jspb.Message.getBooleanFieldWithDefault(msg, 3, false),
    styling: (f = msg.getStyling()) && gendocu_documentation_v1_styling_pb.StylingSettings.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GetDocumentationResponse;
  return proto.gendocu.sdk_generator.v3.GetDocumentationResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.Build;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.Build.deserializeBinaryFromReader);
      msg.setBuild(value);
      break;
    case 2:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setIsPrivate(value);
      break;
    case 3:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setCanEditDocumentation(value);
      break;
    case 4:
      var value = new gendocu_documentation_v1_styling_pb.StylingSettings;
      reader.readMessage(value,gendocu_documentation_v1_styling_pb.StylingSettings.deserializeBinaryFromReader);
      msg.setStyling(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GetDocumentationResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuild();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.Build.serializeBinaryToWriter
    );
  }
  f = message.getIsPrivate();
  if (f) {
    writer.writeBool(
      2,
      f
    );
  }
  f = message.getCanEditDocumentation();
  if (f) {
    writer.writeBool(
      3,
      f
    );
  }
  f = message.getStyling();
  if (f != null) {
    writer.writeMessage(
      4,
      f,
      gendocu_documentation_v1_styling_pb.StylingSettings.serializeBinaryToWriter
    );
  }
};


/**
 * optional Build build = 1;
 * @return {?proto.gendocu.sdk_generator.v3.Build}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.getBuild = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.Build} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.Build, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.Build|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
*/
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.setBuild = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.clearBuild = function() {
  return this.setBuild(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.hasBuild = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional bool is_private = 2;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.getIsPrivate = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 2, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.setIsPrivate = function(value) {
  return jspb.Message.setProto3BooleanField(this, 2, value);
};


/**
 * optional bool can_edit_documentation = 3;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.getCanEditDocumentation = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 3, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.setCanEditDocumentation = function(value) {
  return jspb.Message.setProto3BooleanField(this, 3, value);
};


/**
 * optional gendocu.documentation.v1.StylingSettings styling = 4;
 * @return {?proto.gendocu.documentation.v1.StylingSettings}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.getStyling = function() {
  return /** @type{?proto.gendocu.documentation.v1.StylingSettings} */ (
    jspb.Message.getWrapperField(this, gendocu_documentation_v1_styling_pb.StylingSettings, 4));
};


/**
 * @param {?proto.gendocu.documentation.v1.StylingSettings|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
*/
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.setStyling = function(value) {
  return jspb.Message.setWrapperField(this, 4, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.GetDocumentationResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.clearStyling = function() {
  return this.setStyling(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GetDocumentationResponse.prototype.hasStyling = function() {
  return jspb.Message.getField(this, 4) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.EnterPasswordRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    selector: (f = msg.getSelector()) && proto.gendocu.sdk_generator.v3.BuildSelectorRequest.toObject(includeInstance, f),
    password: jspb.Message.getFieldWithDefault(msg, 2, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.EnterPasswordRequest;
  return proto.gendocu.sdk_generator.v3.EnterPasswordRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.BuildSelectorRequest;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.BuildSelectorRequest.deserializeBinaryFromReader);
      msg.setSelector(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setPassword(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.EnterPasswordRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getSelector();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.BuildSelectorRequest.serializeBinaryToWriter
    );
  }
  f = message.getPassword();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
};


/**
 * optional BuildSelectorRequest selector = 1;
 * @return {?proto.gendocu.sdk_generator.v3.BuildSelectorRequest}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.getSelector = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.BuildSelectorRequest} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.BuildSelectorRequest, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.BuildSelectorRequest|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} returns this
*/
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.setSelector = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} returns this
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.clearSelector = function() {
  return this.setSelector(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.hasSelector = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional string password = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.getPassword = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordRequest} returns this
 */
proto.gendocu.sdk_generator.v3.EnterPasswordRequest.prototype.setPassword = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.EnterPasswordResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    sessionId: jspb.Message.getFieldWithDefault(msg, 1, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.EnterPasswordResponse;
  return proto.gendocu.sdk_generator.v3.EnterPasswordResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setSessionId(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.EnterPasswordResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getSessionId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
};


/**
 * optional string session_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.prototype.getSessionId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.EnterPasswordResponse} returns this
 */
proto.gendocu.sdk_generator.v3.EnterPasswordResponse.prototype.setSessionId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.repeatedFields_ = [2];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    buildId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    methodsList: (f = jspb.Message.getRepeatedField(msg, 2)) == null ? undefined : f,
    sdkProvider: jspb.Message.getFieldWithDefault(msg, 3, 0),
    programmingLang: jspb.Message.getFieldWithDefault(msg, 4, 0),
    selectedEnvironment: jspb.Message.getFieldWithDefault(msg, 6, ""),
    exampleId: jspb.Message.getFieldWithDefault(msg, 7, ""),
    authenticationMethodId: jspb.Message.getFieldWithDefault(msg, 8, ""),
    authenticationToken: jspb.Message.getFieldWithDefault(msg, 9, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest;
  return proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.addMethods(value);
      break;
    case 3:
      var value = /** @type {!proto.gendocu.common.SDKProvider} */ (reader.readEnum());
      msg.setSdkProvider(value);
      break;
    case 4:
      var value = /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (reader.readEnum());
      msg.setProgrammingLang(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setSelectedEnvironment(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setExampleId(value);
      break;
    case 8:
      var value = /** @type {string} */ (reader.readString());
      msg.setAuthenticationMethodId(value);
      break;
    case 9:
      var value = /** @type {string} */ (reader.readString());
      msg.setAuthenticationToken(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuildId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getMethodsList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      2,
      f
    );
  }
  f = message.getSdkProvider();
  if (f !== 0.0) {
    writer.writeEnum(
      3,
      f
    );
  }
  f = message.getProgrammingLang();
  if (f !== 0.0) {
    writer.writeEnum(
      4,
      f
    );
  }
  f = message.getSelectedEnvironment();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getExampleId();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getAuthenticationMethodId();
  if (f.length > 0) {
    writer.writeString(
      8,
      f
    );
  }
  f = message.getAuthenticationToken();
  if (f.length > 0) {
    writer.writeString(
      9,
      f
    );
  }
};


/**
 * optional string build_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setBuildId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * repeated string methods = 2;
 * @return {!Array<string>}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getMethodsList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 2));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setMethodsList = function(value) {
  return jspb.Message.setField(this, 2, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.addMethods = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 2, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.clearMethodsList = function() {
  return this.setMethodsList([]);
};


/**
 * optional gendocu.common.SDKProvider sdk_provider = 3;
 * @return {!proto.gendocu.common.SDKProvider}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getSdkProvider = function() {
  return /** @type {!proto.gendocu.common.SDKProvider} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};


/**
 * @param {!proto.gendocu.common.SDKProvider} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setSdkProvider = function(value) {
  return jspb.Message.setProto3EnumField(this, 3, value);
};


/**
 * optional gendocu.common.ProgrammingLanguage programming_lang = 4;
 * @return {!proto.gendocu.common.ProgrammingLanguage}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getProgrammingLang = function() {
  return /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};


/**
 * @param {!proto.gendocu.common.ProgrammingLanguage} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setProgrammingLang = function(value) {
  return jspb.Message.setProto3EnumField(this, 4, value);
};


/**
 * optional string selected_environment = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getSelectedEnvironment = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setSelectedEnvironment = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string example_id = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getExampleId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setExampleId = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional string authentication_method_id = 8;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getAuthenticationMethodId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 8, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setAuthenticationMethodId = function(value) {
  return jspb.Message.setProto3StringField(this, 8, value);
};


/**
 * optional string authentication_token = 9;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.getAuthenticationToken = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 9, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleRequest.prototype.setAuthenticationToken = function(value) {
  return jspb.Message.setProto3StringField(this, 9, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.repeatedFields_ = [5];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    code: jspb.Message.getFieldWithDefault(msg, 1, ""),
    highlightsMap: (f = msg.getHighlightsMap()) ? f.toObject(includeInstance, undefined) : [],
    methodIdToRichHighlightMap: (f = msg.getMethodIdToRichHighlightMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.Lines.toObject) : [],
    compilationErrorsMap: (f = msg.getCompilationErrorsMap()) ? f.toObject(includeInstance, undefined) : [],
    programmingLang: jspb.Message.getFieldWithDefault(msg, 4, 0),
    setupStepsList: (f = jspb.Message.getRepeatedField(msg, 5)) == null ? undefined : f,
    codeOutput: jspb.Message.getFieldWithDefault(msg, 7, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse;
  return proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setCode(value);
      break;
    case 2:
      var value = msg.getHighlightsMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readInt32, jspb.BinaryReader.prototype.readInt32, null, 0, 0);
         });
      break;
    case 6:
      var value = msg.getMethodIdToRichHighlightMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.Lines.deserializeBinaryFromReader, "", new proto.gendocu.sdk_generator.v3.Lines());
         });
      break;
    case 3:
      var value = msg.getCompilationErrorsMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readInt32, jspb.BinaryReader.prototype.readString, null, 0, "");
         });
      break;
    case 4:
      var value = /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (reader.readEnum());
      msg.setProgrammingLang(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.addSetupSteps(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setCodeOutput(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getCode();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getHighlightsMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(2, writer, jspb.BinaryWriter.prototype.writeInt32, jspb.BinaryWriter.prototype.writeInt32);
  }
  f = message.getMethodIdToRichHighlightMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(6, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.Lines.serializeBinaryToWriter);
  }
  f = message.getCompilationErrorsMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(3, writer, jspb.BinaryWriter.prototype.writeInt32, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getProgrammingLang();
  if (f !== 0.0) {
    writer.writeEnum(
      4,
      f
    );
  }
  f = message.getSetupStepsList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      5,
      f
    );
  }
  f = message.getCodeOutput();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
};


/**
 * optional string code = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getCode = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.setCode = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * map<int32, int32> highlights = 2;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<number,number>}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getHighlightsMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<number,number>} */ (
      jspb.Message.getMapField(this, 2, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.clearHighlightsMap = function() {
  this.getHighlightsMap().clear();
  return this;};


/**
 * map<string, Lines> method_id_to_rich_highlight = 6;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.Lines>}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getMethodIdToRichHighlightMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.Lines>} */ (
      jspb.Message.getMapField(this, 6, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.Lines));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.clearMethodIdToRichHighlightMap = function() {
  this.getMethodIdToRichHighlightMap().clear();
  return this;};


/**
 * map<int32, string> compilation_errors = 3;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<number,string>}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getCompilationErrorsMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<number,string>} */ (
      jspb.Message.getMapField(this, 3, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.clearCompilationErrorsMap = function() {
  this.getCompilationErrorsMap().clear();
  return this;};


/**
 * optional gendocu.common.ProgrammingLanguage programming_lang = 4;
 * @return {!proto.gendocu.common.ProgrammingLanguage}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getProgrammingLang = function() {
  return /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};


/**
 * @param {!proto.gendocu.common.ProgrammingLanguage} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.setProgrammingLang = function(value) {
  return jspb.Message.setProto3EnumField(this, 4, value);
};


/**
 * repeated string setup_steps = 5;
 * @return {!Array<string>}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getSetupStepsList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 5));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.setSetupStepsList = function(value) {
  return jspb.Message.setField(this, 5, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.addSetupSteps = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 5, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.clearSetupStepsList = function() {
  return this.setSetupStepsList([]);
};


/**
 * optional string code_output = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.getCodeOutput = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateCodeSampleResponse.prototype.setCodeOutput = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.Lines.repeatedFields_ = [1];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.Lines.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.Lines} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Lines.toObject = function(includeInstance, msg) {
  var f, obj = {
    lineList: (f = jspb.Message.getRepeatedField(msg, 1)) == null ? undefined : f
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.Lines}
 */
proto.gendocu.sdk_generator.v3.Lines.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.Lines;
  return proto.gendocu.sdk_generator.v3.Lines.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.Lines} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.Lines}
 */
proto.gendocu.sdk_generator.v3.Lines.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var values = /** @type {!Array<number>} */ (reader.isDelimited() ? reader.readPackedInt32() : [reader.readInt32()]);
      for (var i = 0; i < values.length; i++) {
        msg.addLine(values[i]);
      }
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.Lines.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.Lines} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Lines.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getLineList();
  if (f.length > 0) {
    writer.writePackedInt32(
      1,
      f
    );
  }
};


/**
 * repeated int32 line = 1;
 * @return {!Array<number>}
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.getLineList = function() {
  return /** @type {!Array<number>} */ (jspb.Message.getRepeatedField(this, 1));
};


/**
 * @param {!Array<number>} value
 * @return {!proto.gendocu.sdk_generator.v3.Lines} returns this
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.setLineList = function(value) {
  return jspb.Message.setField(this, 1, value || []);
};


/**
 * @param {number} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.Lines} returns this
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.addLine = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 1, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Lines} returns this
 */
proto.gendocu.sdk_generator.v3.Lines.prototype.clearLineList = function() {
  return this.setLineList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    buildId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    sdkProvider: jspb.Message.getFieldWithDefault(msg, 2, 0),
    programmingLang: jspb.Message.getFieldWithDefault(msg, 3, 0),
    authenticationMethodId: jspb.Message.getFieldWithDefault(msg, 4, ""),
    selectedEnvironment: jspb.Message.getFieldWithDefault(msg, 5, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest;
  return proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    case 2:
      var value = /** @type {!proto.gendocu.common.SDKProvider} */ (reader.readEnum());
      msg.setSdkProvider(value);
      break;
    case 3:
      var value = /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (reader.readEnum());
      msg.setProgrammingLang(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setAuthenticationMethodId(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setSelectedEnvironment(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuildId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getSdkProvider();
  if (f !== 0.0) {
    writer.writeEnum(
      2,
      f
    );
  }
  f = message.getProgrammingLang();
  if (f !== 0.0) {
    writer.writeEnum(
      3,
      f
    );
  }
  f = message.getAuthenticationMethodId();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getSelectedEnvironment();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
};


/**
 * optional string build_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.setBuildId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional gendocu.common.SDKProvider sdk_provider = 2;
 * @return {!proto.gendocu.common.SDKProvider}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.getSdkProvider = function() {
  return /** @type {!proto.gendocu.common.SDKProvider} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
};


/**
 * @param {!proto.gendocu.common.SDKProvider} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.setSdkProvider = function(value) {
  return jspb.Message.setProto3EnumField(this, 2, value);
};


/**
 * optional gendocu.common.ProgrammingLanguage programming_lang = 3;
 * @return {!proto.gendocu.common.ProgrammingLanguage}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.getProgrammingLang = function() {
  return /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};


/**
 * @param {!proto.gendocu.common.ProgrammingLanguage} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.setProgrammingLang = function(value) {
  return jspb.Message.setProto3EnumField(this, 3, value);
};


/**
 * optional string authentication_method_id = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.getAuthenticationMethodId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.setAuthenticationMethodId = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional string selected_environment = 5;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.getSelectedEnvironment = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsRequest.prototype.setSelectedEnvironment = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    methodIdToCodeSnippetMap: (f = msg.getMethodIdToCodeSnippetMap()) ? f.toObject(includeInstance, undefined) : [],
    programmingLang: jspb.Message.getFieldWithDefault(msg, 2, 0),
    methodIdToCodeOutputMap: (f = msg.getMethodIdToCodeOutputMap()) ? f.toObject(includeInstance, undefined) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse;
  return proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = msg.getMethodIdToCodeSnippetMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    case 2:
      var value = /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (reader.readEnum());
      msg.setProgrammingLang(value);
      break;
    case 3:
      var value = msg.getMethodIdToCodeOutputMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMethodIdToCodeSnippetMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(1, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getProgrammingLang();
  if (f !== 0.0) {
    writer.writeEnum(
      2,
      f
    );
  }
  f = message.getMethodIdToCodeOutputMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(3, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
};


/**
 * map<string, string> method_id_to_code_snippet = 1;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.getMethodIdToCodeSnippetMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 1, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.clearMethodIdToCodeSnippetMap = function() {
  this.getMethodIdToCodeSnippetMap().clear();
  return this;};


/**
 * optional gendocu.common.ProgrammingLanguage programming_lang = 2;
 * @return {!proto.gendocu.common.ProgrammingLanguage}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.getProgrammingLang = function() {
  return /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
};


/**
 * @param {!proto.gendocu.common.ProgrammingLanguage} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.setProgrammingLang = function(value) {
  return jspb.Message.setProto3EnumField(this, 2, value);
};


/**
 * map<string, string> method_id_to_code_output = 3;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.getMethodIdToCodeOutputMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 3, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateInputCodeSnippetsResponse.prototype.clearMethodIdToCodeOutputMap = function() {
  this.getMethodIdToCodeOutputMap().clear();
  return this;};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.repeatedFields_ = [12];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.CreateAPIDecode.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.toObject = function(includeInstance, msg) {
  var f, obj = {
    buildId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    buildStarted: (f = msg.getBuildStarted()) && google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
    organization: jspb.Message.getFieldWithDefault(msg, 2, ""),
    apiName: jspb.Message.getFieldWithDefault(msg, 3, ""),
    projectId: jspb.Message.getFieldWithDefault(msg, 4, ""),
    description: (f = msg.getDescription()) && gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.toObject(includeInstance, f),
    giteaName: jspb.Message.getFieldWithDefault(msg, 6, ""),
    apiVersion: jspb.Message.getFieldWithDefault(msg, 7, ""),
    authenticationDescription: (f = msg.getAuthenticationDescription()) && gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.toObject(includeInstance, f),
    sdkProvisioning: (f = msg.getSdkProvisioning()) && proto.gendocu.sdk_generator.v3.SDKProvisioning.toObject(includeInstance, f),
    environmentRoutingList: jspb.Message.toObjectList(msg.getEnvironmentRoutingList(),
    proto.gendocu.sdk_generator.v3.EnvironmentsRouting.toObject, includeInstance)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.CreateAPIDecode;
  return proto.gendocu.sdk_generator.v3.CreateAPIDecode.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    case 13:
      var value = new google_protobuf_timestamp_pb.Timestamp;
      reader.readMessage(value,google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader);
      msg.setBuildStarted(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setOrganization(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setApiName(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 5:
      var value = new gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel;
      reader.readMessage(value,gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.deserializeBinaryFromReader);
      msg.setDescription(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setGiteaName(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setApiVersion(value);
      break;
    case 10:
      var value = new gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription;
      reader.readMessage(value,gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.deserializeBinaryFromReader);
      msg.setAuthenticationDescription(value);
      break;
    case 11:
      var value = new proto.gendocu.sdk_generator.v3.SDKProvisioning;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.SDKProvisioning.deserializeBinaryFromReader);
      msg.setSdkProvisioning(value);
      break;
    case 12:
      var value = new proto.gendocu.sdk_generator.v3.EnvironmentsRouting;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.EnvironmentsRouting.deserializeBinaryFromReader);
      msg.addEnvironmentRouting(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.CreateAPIDecode.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuildId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getBuildStarted();
  if (f != null) {
    writer.writeMessage(
      13,
      f,
      google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter
    );
  }
  f = message.getOrganization();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getApiName();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getDescription();
  if (f != null) {
    writer.writeMessage(
      5,
      f,
      gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.serializeBinaryToWriter
    );
  }
  f = message.getGiteaName();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getApiVersion();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getAuthenticationDescription();
  if (f != null) {
    writer.writeMessage(
      10,
      f,
      gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.serializeBinaryToWriter
    );
  }
  f = message.getSdkProvisioning();
  if (f != null) {
    writer.writeMessage(
      11,
      f,
      proto.gendocu.sdk_generator.v3.SDKProvisioning.serializeBinaryToWriter
    );
  }
  f = message.getEnvironmentRoutingList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      12,
      f,
      proto.gendocu.sdk_generator.v3.EnvironmentsRouting.serializeBinaryToWriter
    );
  }
};


/**
 * optional string build_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setBuildId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional google.protobuf.Timestamp build_started = 13;
 * @return {?proto.google.protobuf.Timestamp}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getBuildStarted = function() {
  return /** @type{?proto.google.protobuf.Timestamp} */ (
    jspb.Message.getWrapperField(this, google_protobuf_timestamp_pb.Timestamp, 13));
};


/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
*/
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setBuildStarted = function(value) {
  return jspb.Message.setWrapperField(this, 13, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.clearBuildStarted = function() {
  return this.setBuildStarted(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.hasBuildStarted = function() {
  return jspb.Message.getField(this, 13) != null;
};


/**
 * optional string organization = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getOrganization = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setOrganization = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string api_name = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getApiName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setApiName = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string project_id = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional APIDescriptionGeneratorModel description = 5;
 * @return {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getDescription = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} */ (
    jspb.Message.getWrapperField(this, gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel, 5));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
*/
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setDescription = function(value) {
  return jspb.Message.setWrapperField(this, 5, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.clearDescription = function() {
  return this.setDescription(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.hasDescription = function() {
  return jspb.Message.getField(this, 5) != null;
};


/**
 * optional string gitea_name = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getGiteaName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setGiteaName = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string api_version = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getApiVersion = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setApiVersion = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional AuthenticationDescription authentication_description = 10;
 * @return {?proto.gendocu.sdk_generator.v3.AuthenticationDescription}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getAuthenticationDescription = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.AuthenticationDescription} */ (
    jspb.Message.getWrapperField(this, gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription, 10));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.AuthenticationDescription|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
*/
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setAuthenticationDescription = function(value) {
  return jspb.Message.setWrapperField(this, 10, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.clearAuthenticationDescription = function() {
  return this.setAuthenticationDescription(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.hasAuthenticationDescription = function() {
  return jspb.Message.getField(this, 10) != null;
};


/**
 * optional SDKProvisioning sdk_provisioning = 11;
 * @return {?proto.gendocu.sdk_generator.v3.SDKProvisioning}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getSdkProvisioning = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.SDKProvisioning} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.SDKProvisioning, 11));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.SDKProvisioning|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
*/
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setSdkProvisioning = function(value) {
  return jspb.Message.setWrapperField(this, 11, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.clearSdkProvisioning = function() {
  return this.setSdkProvisioning(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.hasSdkProvisioning = function() {
  return jspb.Message.getField(this, 11) != null;
};


/**
 * repeated EnvironmentsRouting environment_routing = 12;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.EnvironmentsRouting>}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.getEnvironmentRoutingList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.EnvironmentsRouting>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.EnvironmentsRouting, 12));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.EnvironmentsRouting>} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
*/
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.setEnvironmentRoutingList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 12, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.addEnvironmentRouting = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 12, opt_value, proto.gendocu.sdk_generator.v3.EnvironmentsRouting, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecode} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecode.prototype.clearEnvironmentRoutingList = function() {
  return this.setEnvironmentRoutingList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.EnvironmentsRouting.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.toObject = function(includeInstance, msg) {
  var f, obj = {
    environmentName: jspb.Message.getFieldWithDefault(msg, 1, ""),
    serviceToUrlsMap: (f = msg.getServiceToUrlsMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.ServiceRouting.toObject) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.EnvironmentsRouting;
  return proto.gendocu.sdk_generator.v3.EnvironmentsRouting.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setEnvironmentName(value);
      break;
    case 2:
      var value = msg.getServiceToUrlsMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.ServiceRouting.deserializeBinaryFromReader, "", new proto.gendocu.sdk_generator.v3.ServiceRouting());
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.EnvironmentsRouting.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getEnvironmentName();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getServiceToUrlsMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(2, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.ServiceRouting.serializeBinaryToWriter);
  }
};


/**
 * optional string environment_name = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.getEnvironmentName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting} returns this
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.setEnvironmentName = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * map<string, ServiceRouting> service_to_urls = 2;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.ServiceRouting>}
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.getServiceToUrlsMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.ServiceRouting>} */ (
      jspb.Message.getMapField(this, 2, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.ServiceRouting));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.EnvironmentsRouting} returns this
 */
proto.gendocu.sdk_generator.v3.EnvironmentsRouting.prototype.clearServiceToUrlsMap = function() {
  this.getServiceToUrlsMap().clear();
  return this;};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.repeatedFields_ = [2];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.ServiceRouting.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.ServiceRouting} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.toObject = function(includeInstance, msg) {
  var f, obj = {
    urlList: (f = jspb.Message.getRepeatedField(msg, 2)) == null ? undefined : f
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceRouting}
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.ServiceRouting;
  return proto.gendocu.sdk_generator.v3.ServiceRouting.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.ServiceRouting} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceRouting}
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.addUrl(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.ServiceRouting.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.ServiceRouting} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getUrlList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      2,
      f
    );
  }
};


/**
 * repeated string url = 2;
 * @return {!Array<string>}
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.getUrlList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 2));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceRouting} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.setUrlList = function(value) {
  return jspb.Message.setField(this, 2, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.ServiceRouting} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.addUrl = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 2, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceRouting} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceRouting.prototype.clearUrlList = function() {
  return this.setUrlList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    apiUrl: jspb.Message.getFieldWithDefault(msg, 1, ""),
    buildId: jspb.Message.getFieldWithDefault(msg, 2, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse;
  return proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setApiUrl(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getApiUrl();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getBuildId();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
};


/**
 * optional string api_url = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.getApiUrl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.setApiUrl = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string build_id = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse} returns this
 */
proto.gendocu.sdk_generator.v3.CreateAPIDecodeResponse.prototype.setBuildId = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.repeatedFields_ = [1];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.GenerateSDKResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    methodsList: (f = jspb.Message.getRepeatedField(msg, 1)) == null ? undefined : f,
    apiDescription: (f = msg.getApiDescription()) && gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.toObject(includeInstance, f),
    sdkProvider: jspb.Message.getFieldWithDefault(msg, 3, 0),
    lang: jspb.Message.getFieldWithDefault(msg, 4, 0)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.GenerateSDKResponse;
  return proto.gendocu.sdk_generator.v3.GenerateSDKResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.addMethods(value);
      break;
    case 2:
      var value = new gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel;
      reader.readMessage(value,gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.deserializeBinaryFromReader);
      msg.setApiDescription(value);
      break;
    case 3:
      var value = /** @type {!proto.gendocu.common.SDKProvider} */ (reader.readEnum());
      msg.setSdkProvider(value);
      break;
    case 4:
      var value = /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (reader.readEnum());
      msg.setLang(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.GenerateSDKResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMethodsList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      1,
      f
    );
  }
  f = message.getApiDescription();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.serializeBinaryToWriter
    );
  }
  f = message.getSdkProvider();
  if (f !== 0.0) {
    writer.writeEnum(
      3,
      f
    );
  }
  f = message.getLang();
  if (f !== 0.0) {
    writer.writeEnum(
      4,
      f
    );
  }
};


/**
 * repeated string methods = 1;
 * @return {!Array<string>}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.getMethodsList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 1));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.setMethodsList = function(value) {
  return jspb.Message.setField(this, 1, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.addMethods = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 1, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.clearMethodsList = function() {
  return this.setMethodsList([]);
};


/**
 * optional APIDescriptionGeneratorModel api_description = 2;
 * @return {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.getApiDescription = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} */ (
    jspb.Message.getWrapperField(this, gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel, 2));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
*/
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.setApiDescription = function(value) {
  return jspb.Message.setWrapperField(this, 2, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.clearApiDescription = function() {
  return this.setApiDescription(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.hasApiDescription = function() {
  return jspb.Message.getField(this, 2) != null;
};


/**
 * optional gendocu.common.SDKProvider sdk_provider = 3;
 * @return {!proto.gendocu.common.SDKProvider}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.getSdkProvider = function() {
  return /** @type {!proto.gendocu.common.SDKProvider} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};


/**
 * @param {!proto.gendocu.common.SDKProvider} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.setSdkProvider = function(value) {
  return jspb.Message.setProto3EnumField(this, 3, value);
};


/**
 * optional gendocu.common.ProgrammingLanguage lang = 4;
 * @return {!proto.gendocu.common.ProgrammingLanguage}
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.getLang = function() {
  return /** @type {!proto.gendocu.common.ProgrammingLanguage} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};


/**
 * @param {!proto.gendocu.common.ProgrammingLanguage} value
 * @return {!proto.gendocu.sdk_generator.v3.GenerateSDKResponse} returns this
 */
proto.gendocu.sdk_generator.v3.GenerateSDKResponse.prototype.setLang = function(value) {
  return jspb.Message.setProto3EnumField(this, 4, value);
};



/**
 * Oneof group definitions for this message. Each group defines the field
 * numbers belonging to that group. When of these fields' value is set, all
 * other fields in the group are cleared. During deserialization, if multiple
 * fields are encountered for a group, only the last value seen will be kept.
 * @private {!Array<!Array<number>>}
 * @const
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_ = [[1,2]];

/**
 * @enum {number}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.FilterCase = {
  FILTER_NOT_SET: 0,
  BUILD_ID: 1,
  ORG_AND_NAME: 2
};

/**
 * @return {proto.gendocu.sdk_generator.v3.BuildSelectorRequest.FilterCase}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.getFilterCase = function() {
  return /** @type {proto.gendocu.sdk_generator.v3.BuildSelectorRequest.FilterCase} */(jspb.Message.computeOneofCase(this, proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_[0]));
};



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.BuildSelectorRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    buildId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    orgAndName: (f = msg.getOrgAndName()) && proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.BuildSelectorRequest;
  return proto.gendocu.sdk_generator.v3.BuildSelectorRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    case 2:
      var value = new proto.gendocu.sdk_generator.v3.BuildWithOrgAndName;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.deserializeBinaryFromReader);
      msg.setOrgAndName(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.BuildSelectorRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = /** @type {string} */ (jspb.Message.getField(message, 1));
  if (f != null) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getOrgAndName();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.serializeBinaryToWriter
    );
  }
};


/**
 * optional string build_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} returns this
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.setBuildId = function(value) {
  return jspb.Message.setOneofField(this, 1, proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_[0], value);
};


/**
 * Clears the field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} returns this
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.clearBuildId = function() {
  return jspb.Message.setOneofField(this, 1, proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_[0], undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.hasBuildId = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional BuildWithOrgAndName org_and_name = 2;
 * @return {?proto.gendocu.sdk_generator.v3.BuildWithOrgAndName}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.getOrgAndName = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.BuildWithOrgAndName, 2));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.BuildWithOrgAndName|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} returns this
*/
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.setOrgAndName = function(value) {
  return jspb.Message.setOneofWrapperField(this, 2, proto.gendocu.sdk_generator.v3.BuildSelectorRequest.oneofGroups_[0], value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.BuildSelectorRequest} returns this
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.clearOrgAndName = function() {
  return this.setOrgAndName(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.BuildSelectorRequest.prototype.hasOrgAndName = function() {
  return jspb.Message.getField(this, 2) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.toObject = function(includeInstance, msg) {
  var f, obj = {
    apiName: jspb.Message.getFieldWithDefault(msg, 1, ""),
    organization: jspb.Message.getFieldWithDefault(msg, 2, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.BuildWithOrgAndName;
  return proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setApiName(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setOrganization(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getApiName();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getOrganization();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
};


/**
 * optional string api_name = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.getApiName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} returns this
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.setApiName = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string organization = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.getOrganization = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.BuildWithOrgAndName} returns this
 */
proto.gendocu.sdk_generator.v3.BuildWithOrgAndName.prototype.setOrganization = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.Build.repeatedFields_ = [8];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.Build.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.Build} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Build.toObject = function(includeInstance, msg) {
  var f, obj = {
    buildId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    organization: jspb.Message.getFieldWithDefault(msg, 2, ""),
    apiName: jspb.Message.getFieldWithDefault(msg, 3, ""),
    data: (f = msg.getData()) && gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.toObject(includeInstance, f),
    sdkProvisioning: (f = msg.getSdkProvisioning()) && proto.gendocu.sdk_generator.v3.SDKProvisioning.toObject(includeInstance, f),
    authenticationDescription: (f = msg.getAuthenticationDescription()) && gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.toObject(includeInstance, f),
    envsList: jspb.Message.toObjectList(msg.getEnvsList(),
    proto.gendocu.sdk_generator.v3.Environment.toObject, includeInstance),
    defaultEnvironment: jspb.Message.getFieldWithDefault(msg, 9, ""),
    grpcRepoSourceUrl: jspb.Message.getFieldWithDefault(msg, 10, ""),
    swaggerFile: jspb.Message.getFieldWithDefault(msg, 12, ""),
    tryItEnabled: jspb.Message.getBooleanFieldWithDefault(msg, 11, false),
    serviceGrpcUrlMap: (f = msg.getServiceGrpcUrlMap()) ? f.toObject(includeInstance, undefined) : [],
    serviceWebgrpcUrlMap: (f = msg.getServiceWebgrpcUrlMap()) ? f.toObject(includeInstance, undefined) : [],
    projectId: jspb.Message.getFieldWithDefault(msg, 24, ""),
    sdkBranch: jspb.Message.getFieldWithDefault(msg, 25, ""),
    logoUrl: jspb.Message.getFieldWithDefault(msg, 40, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.Build}
 */
proto.gendocu.sdk_generator.v3.Build.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.Build;
  return proto.gendocu.sdk_generator.v3.Build.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.Build} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.Build}
 */
proto.gendocu.sdk_generator.v3.Build.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setBuildId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setOrganization(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setApiName(value);
      break;
    case 5:
      var value = new gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel;
      reader.readMessage(value,gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.deserializeBinaryFromReader);
      msg.setData(value);
      break;
    case 6:
      var value = new proto.gendocu.sdk_generator.v3.SDKProvisioning;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.SDKProvisioning.deserializeBinaryFromReader);
      msg.setSdkProvisioning(value);
      break;
    case 7:
      var value = new gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription;
      reader.readMessage(value,gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.deserializeBinaryFromReader);
      msg.setAuthenticationDescription(value);
      break;
    case 8:
      var value = new proto.gendocu.sdk_generator.v3.Environment;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.Environment.deserializeBinaryFromReader);
      msg.addEnvs(value);
      break;
    case 9:
      var value = /** @type {string} */ (reader.readString());
      msg.setDefaultEnvironment(value);
      break;
    case 10:
      var value = /** @type {string} */ (reader.readString());
      msg.setGrpcRepoSourceUrl(value);
      break;
    case 12:
      var value = /** @type {string} */ (reader.readString());
      msg.setSwaggerFile(value);
      break;
    case 11:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setTryItEnabled(value);
      break;
    case 22:
      var value = msg.getServiceGrpcUrlMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    case 23:
      var value = msg.getServiceWebgrpcUrlMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    case 24:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 25:
      var value = /** @type {string} */ (reader.readString());
      msg.setSdkBranch(value);
      break;
    case 40:
      var value = /** @type {string} */ (reader.readString());
      msg.setLogoUrl(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.Build.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.Build} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Build.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuildId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getOrganization();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getApiName();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getData();
  if (f != null) {
    writer.writeMessage(
      5,
      f,
      gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel.serializeBinaryToWriter
    );
  }
  f = message.getSdkProvisioning();
  if (f != null) {
    writer.writeMessage(
      6,
      f,
      proto.gendocu.sdk_generator.v3.SDKProvisioning.serializeBinaryToWriter
    );
  }
  f = message.getAuthenticationDescription();
  if (f != null) {
    writer.writeMessage(
      7,
      f,
      gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription.serializeBinaryToWriter
    );
  }
  f = message.getEnvsList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      8,
      f,
      proto.gendocu.sdk_generator.v3.Environment.serializeBinaryToWriter
    );
  }
  f = message.getDefaultEnvironment();
  if (f.length > 0) {
    writer.writeString(
      9,
      f
    );
  }
  f = message.getGrpcRepoSourceUrl();
  if (f.length > 0) {
    writer.writeString(
      10,
      f
    );
  }
  f = message.getSwaggerFile();
  if (f.length > 0) {
    writer.writeString(
      12,
      f
    );
  }
  f = message.getTryItEnabled();
  if (f) {
    writer.writeBool(
      11,
      f
    );
  }
  f = message.getServiceGrpcUrlMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(22, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getServiceWebgrpcUrlMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(23, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      24,
      f
    );
  }
  f = message.getSdkBranch();
  if (f.length > 0) {
    writer.writeString(
      25,
      f
    );
  }
  f = message.getLogoUrl();
  if (f.length > 0) {
    writer.writeString(
      40,
      f
    );
  }
};


/**
 * optional string build_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getBuildId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setBuildId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string organization = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getOrganization = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setOrganization = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string api_name = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getApiName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setApiName = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional APIDescriptionGeneratorModel data = 5;
 * @return {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getData = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} */ (
    jspb.Message.getWrapperField(this, gendocu_sdk_generator_v3_sdk_generator_pb.APIDescriptionGeneratorModel, 5));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
*/
proto.gendocu.sdk_generator.v3.Build.prototype.setData = function(value) {
  return jspb.Message.setWrapperField(this, 5, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearData = function() {
  return this.setData(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.hasData = function() {
  return jspb.Message.getField(this, 5) != null;
};


/**
 * optional SDKProvisioning sdk_provisioning = 6;
 * @return {?proto.gendocu.sdk_generator.v3.SDKProvisioning}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getSdkProvisioning = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.SDKProvisioning} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.SDKProvisioning, 6));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.SDKProvisioning|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
*/
proto.gendocu.sdk_generator.v3.Build.prototype.setSdkProvisioning = function(value) {
  return jspb.Message.setWrapperField(this, 6, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearSdkProvisioning = function() {
  return this.setSdkProvisioning(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.hasSdkProvisioning = function() {
  return jspb.Message.getField(this, 6) != null;
};


/**
 * optional AuthenticationDescription authentication_description = 7;
 * @return {?proto.gendocu.sdk_generator.v3.AuthenticationDescription}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getAuthenticationDescription = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.AuthenticationDescription} */ (
    jspb.Message.getWrapperField(this, gendocu_sdk_generator_v3_authentication_model_pb.AuthenticationDescription, 7));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.AuthenticationDescription|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
*/
proto.gendocu.sdk_generator.v3.Build.prototype.setAuthenticationDescription = function(value) {
  return jspb.Message.setWrapperField(this, 7, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearAuthenticationDescription = function() {
  return this.setAuthenticationDescription(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.hasAuthenticationDescription = function() {
  return jspb.Message.getField(this, 7) != null;
};


/**
 * repeated Environment envs = 8;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.Environment>}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getEnvsList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.Environment>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.Environment, 8));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.Environment>} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
*/
proto.gendocu.sdk_generator.v3.Build.prototype.setEnvsList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 8, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.Environment=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.Environment}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.addEnvs = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 8, opt_value, proto.gendocu.sdk_generator.v3.Environment, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearEnvsList = function() {
  return this.setEnvsList([]);
};


/**
 * optional string default_environment = 9;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getDefaultEnvironment = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 9, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setDefaultEnvironment = function(value) {
  return jspb.Message.setProto3StringField(this, 9, value);
};


/**
 * optional string grpc_repo_source_url = 10;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getGrpcRepoSourceUrl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 10, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setGrpcRepoSourceUrl = function(value) {
  return jspb.Message.setProto3StringField(this, 10, value);
};


/**
 * optional string swagger_file = 12;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getSwaggerFile = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 12, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setSwaggerFile = function(value) {
  return jspb.Message.setProto3StringField(this, 12, value);
};


/**
 * optional bool try_it_enabled = 11;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getTryItEnabled = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 11, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setTryItEnabled = function(value) {
  return jspb.Message.setProto3BooleanField(this, 11, value);
};


/**
 * map<string, string> service_grpc_url = 22;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getServiceGrpcUrlMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 22, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearServiceGrpcUrlMap = function() {
  this.getServiceGrpcUrlMap().clear();
  return this;};


/**
 * map<string, string> service_webgrpc_url = 23;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getServiceWebgrpcUrlMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 23, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.clearServiceWebgrpcUrlMap = function() {
  this.getServiceWebgrpcUrlMap().clear();
  return this;};


/**
 * optional string project_id = 24;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 24, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 24, value);
};


/**
 * optional string sdk_branch = 25;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getSdkBranch = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 25, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setSdkBranch = function(value) {
  return jspb.Message.setProto3StringField(this, 25, value);
};


/**
 * optional string logo_url = 40;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Build.prototype.getLogoUrl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 40, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Build} returns this
 */
proto.gendocu.sdk_generator.v3.Build.prototype.setLogoUrl = function(value) {
  return jspb.Message.setProto3StringField(this, 40, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.Environment.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.Environment} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Environment.toObject = function(includeInstance, msg) {
  var f, obj = {
    id: jspb.Message.getFieldWithDefault(msg, 1, ""),
    serviceGrpcUrlMap: (f = msg.getServiceGrpcUrlMap()) ? f.toObject(includeInstance, undefined) : [],
    serviceWebgrpcUrlMap: (f = msg.getServiceWebgrpcUrlMap()) ? f.toObject(includeInstance, undefined) : [],
    serviceProxyUrlMap: (f = msg.getServiceProxyUrlMap()) ? f.toObject(includeInstance, undefined) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.Environment}
 */
proto.gendocu.sdk_generator.v3.Environment.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.Environment;
  return proto.gendocu.sdk_generator.v3.Environment.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.Environment} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.Environment}
 */
proto.gendocu.sdk_generator.v3.Environment.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setId(value);
      break;
    case 2:
      var value = msg.getServiceGrpcUrlMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    case 3:
      var value = msg.getServiceWebgrpcUrlMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    case 4:
      var value = msg.getServiceProxyUrlMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.Environment.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.Environment} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Environment.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getServiceGrpcUrlMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(2, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getServiceWebgrpcUrlMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(3, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
  f = message.getServiceProxyUrlMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(4, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
};


/**
 * optional string id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.getId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Environment} returns this
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.setId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * map<string, string> service_grpc_url = 2;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.getServiceGrpcUrlMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 2, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Environment} returns this
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.clearServiceGrpcUrlMap = function() {
  this.getServiceGrpcUrlMap().clear();
  return this;};


/**
 * map<string, string> service_webgrpc_url = 3;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.getServiceWebgrpcUrlMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 3, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Environment} returns this
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.clearServiceWebgrpcUrlMap = function() {
  this.getServiceWebgrpcUrlMap().clear();
  return this;};


/**
 * map<string, string> service_proxy_url = 4;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.getServiceProxyUrlMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 4, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.Environment} returns this
 */
proto.gendocu.sdk_generator.v3.Environment.prototype.clearServiceProxyUrlMap = function() {
  this.getServiceProxyUrlMap().clear();
  return this;};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.SDKProvisioning.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.SDKProvisioning} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.toObject = function(includeInstance, msg) {
  var f, obj = {
    fetchUrl: jspb.Message.getFieldWithDefault(msg, 1, ""),
    hash: jspb.Message.getFieldWithDefault(msg, 2, ""),
    date: (f = msg.getDate()) && google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
    sdkRepoCommitSha: jspb.Message.getFieldWithDefault(msg, 4, ""),
    sourceCommitSha: jspb.Message.getFieldWithDefault(msg, 5, ""),
    repositoryName: jspb.Message.getFieldWithDefault(msg, 7, ""),
    provisioningDataMap: (f = msg.getProvisioningDataMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.ProvisioningInfo.toObject) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.SDKProvisioning;
  return proto.gendocu.sdk_generator.v3.SDKProvisioning.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.SDKProvisioning} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setFetchUrl(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setHash(value);
      break;
    case 3:
      var value = new google_protobuf_timestamp_pb.Timestamp;
      reader.readMessage(value,google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader);
      msg.setDate(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setSdkRepoCommitSha(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setSourceCommitSha(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setRepositoryName(value);
      break;
    case 6:
      var value = msg.getProvisioningDataMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readInt32, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.ProvisioningInfo.deserializeBinaryFromReader, 0, new proto.gendocu.sdk_generator.v3.ProvisioningInfo());
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.SDKProvisioning.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.SDKProvisioning} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getFetchUrl();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getHash();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getDate();
  if (f != null) {
    writer.writeMessage(
      3,
      f,
      google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter
    );
  }
  f = message.getSdkRepoCommitSha();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getSourceCommitSha();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
  f = message.getRepositoryName();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getProvisioningDataMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(6, writer, jspb.BinaryWriter.prototype.writeInt32, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.ProvisioningInfo.serializeBinaryToWriter);
  }
};


/**
 * optional string fetch_url = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getFetchUrl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setFetchUrl = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string hash = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getHash = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setHash = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional google.protobuf.Timestamp date = 3;
 * @return {?proto.google.protobuf.Timestamp}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getDate = function() {
  return /** @type{?proto.google.protobuf.Timestamp} */ (
    jspb.Message.getWrapperField(this, google_protobuf_timestamp_pb.Timestamp, 3));
};


/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
*/
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setDate = function(value) {
  return jspb.Message.setWrapperField(this, 3, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.clearDate = function() {
  return this.setDate(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.hasDate = function() {
  return jspb.Message.getField(this, 3) != null;
};


/**
 * optional string sdk_repo_commit_sha = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getSdkRepoCommitSha = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setSdkRepoCommitSha = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional string source_commit_sha = 5;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getSourceCommitSha = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setSourceCommitSha = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};


/**
 * optional string repository_name = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getRepositoryName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.setRepositoryName = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * map<int32, ProvisioningInfo> provisioning_data = 6;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<number,!proto.gendocu.sdk_generator.v3.ProvisioningInfo>}
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.getProvisioningDataMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<number,!proto.gendocu.sdk_generator.v3.ProvisioningInfo>} */ (
      jspb.Message.getMapField(this, 6, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.ProvisioningInfo));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.SDKProvisioning} returns this
 */
proto.gendocu.sdk_generator.v3.SDKProvisioning.prototype.clearProvisioningDataMap = function() {
  this.getProvisioningDataMap().clear();
  return this;};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.ProvisioningInfo.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.toObject = function(includeInstance, msg) {
  var f, obj = {
    repositoryUrl: jspb.Message.getFieldWithDefault(msg, 1, ""),
    error: jspb.Message.getFieldWithDefault(msg, 2, ""),
    codeSnippetsSupported: jspb.Message.getBooleanFieldWithDefault(msg, 3, false)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.ProvisioningInfo}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.ProvisioningInfo;
  return proto.gendocu.sdk_generator.v3.ProvisioningInfo.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.ProvisioningInfo}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setRepositoryUrl(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setError(value);
      break;
    case 3:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setCodeSnippetsSupported(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.ProvisioningInfo.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getRepositoryUrl();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getError();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getCodeSnippetsSupported();
  if (f) {
    writer.writeBool(
      3,
      f
    );
  }
};


/**
 * optional string repository_url = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.getRepositoryUrl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} returns this
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.setRepositoryUrl = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string error = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.getError = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} returns this
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.setError = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional bool code_snippets_supported = 3;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.getCodeSnippetsSupported = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 3, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.ProvisioningInfo} returns this
 */
proto.gendocu.sdk_generator.v3.ProvisioningInfo.prototype.setCodeSnippetsSupported = function(value) {
  return jspb.Message.setProto3BooleanField(this, 3, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.UpdateBuildRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    build: (f = msg.getBuild()) && proto.gendocu.sdk_generator.v3.Build.toObject(includeInstance, f),
    updateMask: (f = msg.getUpdateMask()) && google_protobuf_field_mask_pb.FieldMask.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.UpdateBuildRequest;
  return proto.gendocu.sdk_generator.v3.UpdateBuildRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.Build;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.Build.deserializeBinaryFromReader);
      msg.setBuild(value);
      break;
    case 4:
      var value = new google_protobuf_field_mask_pb.FieldMask;
      reader.readMessage(value,google_protobuf_field_mask_pb.FieldMask.deserializeBinaryFromReader);
      msg.setUpdateMask(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.UpdateBuildRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getBuild();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.Build.serializeBinaryToWriter
    );
  }
  f = message.getUpdateMask();
  if (f != null) {
    writer.writeMessage(
      4,
      f,
      google_protobuf_field_mask_pb.FieldMask.serializeBinaryToWriter
    );
  }
};


/**
 * optional Build build = 1;
 * @return {?proto.gendocu.sdk_generator.v3.Build}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.getBuild = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.Build} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.Build, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.Build|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} returns this
*/
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.setBuild = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} returns this
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.clearBuild = function() {
  return this.setBuild(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.hasBuild = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional google.protobuf.FieldMask update_mask = 4;
 * @return {?proto.google.protobuf.FieldMask}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.getUpdateMask = function() {
  return /** @type{?proto.google.protobuf.FieldMask} */ (
    jspb.Message.getWrapperField(this, google_protobuf_field_mask_pb.FieldMask, 4));
};


/**
 * @param {?proto.google.protobuf.FieldMask|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} returns this
*/
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.setUpdateMask = function(value) {
  return jspb.Message.setWrapperField(this, 4, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.UpdateBuildRequest} returns this
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.clearUpdateMask = function() {
  return this.setUpdateMask(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.UpdateBuildRequest.prototype.hasUpdateMask = function() {
  return jspb.Message.getField(this, 4) != null;
};


goog.object.extend(exports, proto.gendocu.sdk_generator.v3);
