// GendocuPublicApis/sdk/ts/gendocu/sdk_generator/v3/authentication_model_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/sdk_generator/v3/authentication_model.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

goog.exportSymbol('proto.gendocu.sdk_generator.v3.AuthenticationDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.AuthenticationMethod', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.CallAuthenticationMethod', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.SchemeCase', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.SchemeCase', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.NoAuthScheme', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.TokenScheme', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.AuthenticationDescription.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.AuthenticationDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.AuthenticationDescription.displayName = 'proto.gendocu.sdk_generator.v3.AuthenticationDescription';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.AuthenticationMethod, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.AuthenticationMethod.displayName = 'proto.gendocu.sdk_generator.v3.AuthenticationMethod';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.oneofGroups_);
};
goog.inherits(proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.displayName = 'proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.oneofGroups_);
};
goog.inherits(proto.gendocu.sdk_generator.v3.CallAuthenticationMethod, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.displayName = 'proto.gendocu.sdk_generator.v3.CallAuthenticationMethod';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.TokenScheme = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.TokenScheme, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.TokenScheme.displayName = 'proto.gendocu.sdk_generator.v3.TokenScheme';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.displayName = 'proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.displayName = 'proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.displayName = 'proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.NoAuthScheme, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.NoAuthScheme.displayName = 'proto.gendocu.sdk_generator.v3.NoAuthScheme';
}

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.repeatedFields_ = [2];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.AuthenticationDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    methodsList: jspb.Message.toObjectList(msg.getMethodsList(),
    proto.gendocu.sdk_generator.v3.AuthenticationMethod.toObject, includeInstance)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationDescription}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.AuthenticationDescription;
  return proto.gendocu.sdk_generator.v3.AuthenticationDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationDescription}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 2:
      var value = new proto.gendocu.sdk_generator.v3.AuthenticationMethod;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.AuthenticationMethod.deserializeBinaryFromReader);
      msg.addMethods(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.AuthenticationDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMethodsList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      2,
      f,
      proto.gendocu.sdk_generator.v3.AuthenticationMethod.serializeBinaryToWriter
    );
  }
};


/**
 * repeated AuthenticationMethod methods = 2;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.AuthenticationMethod>}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.getMethodsList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.AuthenticationMethod>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.AuthenticationMethod, 2));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.AuthenticationMethod>} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationDescription} returns this
*/
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.setMethodsList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 2, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationMethod=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.addMethods = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 2, opt_value, proto.gendocu.sdk_generator.v3.AuthenticationMethod, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationDescription} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationDescription.prototype.clearMethodsList = function() {
  return this.setMethodsList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.AuthenticationMethod.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.toObject = function(includeInstance, msg) {
  var f, obj = {
    id: jspb.Message.getFieldWithDefault(msg, 6, ""),
    title: jspb.Message.getFieldWithDefault(msg, 7, ""),
    description: jspb.Message.getFieldWithDefault(msg, 5, ""),
    callAuthentication: (f = msg.getCallAuthentication()) && proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.toObject(includeInstance, f),
    channelAuthentication: (f = msg.getChannelAuthentication()) && proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.AuthenticationMethod;
  return proto.gendocu.sdk_generator.v3.AuthenticationMethod.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setId(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setTitle(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 8:
      var value = new proto.gendocu.sdk_generator.v3.CallAuthenticationMethod;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.deserializeBinaryFromReader);
      msg.setCallAuthentication(value);
      break;
    case 9:
      var value = new proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.deserializeBinaryFromReader);
      msg.setChannelAuthentication(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.AuthenticationMethod.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getId();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getTitle();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
  f = message.getCallAuthentication();
  if (f != null) {
    writer.writeMessage(
      8,
      f,
      proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.serializeBinaryToWriter
    );
  }
  f = message.getChannelAuthentication();
  if (f != null) {
    writer.writeMessage(
      9,
      f,
      proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.serializeBinaryToWriter
    );
  }
};


/**
 * optional string id = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.getId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.setId = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string title = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.getTitle = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.setTitle = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional string description = 5;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};


/**
 * optional CallAuthenticationMethod call_authentication = 8;
 * @return {?proto.gendocu.sdk_generator.v3.CallAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.getCallAuthentication = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.CallAuthenticationMethod, 8));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.CallAuthenticationMethod|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
*/
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.setCallAuthentication = function(value) {
  return jspb.Message.setWrapperField(this, 8, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.clearCallAuthentication = function() {
  return this.setCallAuthentication(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.hasCallAuthentication = function() {
  return jspb.Message.getField(this, 8) != null;
};


/**
 * optional ChannelAuthenticationMethod channel_authentication = 9;
 * @return {?proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.getChannelAuthentication = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod, 9));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
*/
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.setChannelAuthentication = function(value) {
  return jspb.Message.setWrapperField(this, 9, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.AuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.clearChannelAuthentication = function() {
  return this.setChannelAuthentication(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.AuthenticationMethod.prototype.hasChannelAuthentication = function() {
  return jspb.Message.getField(this, 9) != null;
};



/**
 * Oneof group definitions for this message. Each group defines the field
 * numbers belonging to that group. When of these fields' value is set, all
 * other fields in the group are cleared. During deserialization, if multiple
 * fields are encountered for a group, only the last value seen will be kept.
 * @private {!Array<!Array<number>>}
 * @const
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.oneofGroups_ = [[1]];

/**
 * @enum {number}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.SchemeCase = {
  SCHEME_NOT_SET: 0,
  NO_AUTH: 1
};

/**
 * @return {proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.SchemeCase}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.getSchemeCase = function() {
  return /** @type {proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.SchemeCase} */(jspb.Message.computeOneofCase(this, proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.oneofGroups_[0]));
};



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.toObject = function(includeInstance, msg) {
  var f, obj = {
    noAuth: (f = msg.getNoAuth()) && proto.gendocu.sdk_generator.v3.NoAuthScheme.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod;
  return proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.NoAuthScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.NoAuthScheme.deserializeBinaryFromReader);
      msg.setNoAuth(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getNoAuth();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.NoAuthScheme.serializeBinaryToWriter
    );
  }
};


/**
 * optional NoAuthScheme no_auth = 1;
 * @return {?proto.gendocu.sdk_generator.v3.NoAuthScheme}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.getNoAuth = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.NoAuthScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.NoAuthScheme, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.NoAuthScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} returns this
*/
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.setNoAuth = function(value) {
  return jspb.Message.setOneofWrapperField(this, 1, proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.oneofGroups_[0], value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.clearNoAuth = function() {
  return this.setNoAuth(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.ChannelAuthenticationMethod.prototype.hasNoAuth = function() {
  return jspb.Message.getField(this, 1) != null;
};



/**
 * Oneof group definitions for this message. Each group defines the field
 * numbers belonging to that group. When of these fields' value is set, all
 * other fields in the group are cleared. During deserialization, if multiple
 * fields are encountered for a group, only the last value seen will be kept.
 * @private {!Array<!Array<number>>}
 * @const
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.oneofGroups_ = [[1,2]];

/**
 * @enum {number}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.SchemeCase = {
  SCHEME_NOT_SET: 0,
  NO_AUTH: 1,
  TOKEN_SCHEME: 2
};

/**
 * @return {proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.SchemeCase}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.getSchemeCase = function() {
  return /** @type {proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.SchemeCase} */(jspb.Message.computeOneofCase(this, proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.oneofGroups_[0]));
};



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.toObject = function(includeInstance, msg) {
  var f, obj = {
    noAuth: (f = msg.getNoAuth()) && proto.gendocu.sdk_generator.v3.NoAuthScheme.toObject(includeInstance, f),
    tokenScheme: (f = msg.getTokenScheme()) && proto.gendocu.sdk_generator.v3.TokenScheme.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.CallAuthenticationMethod;
  return proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.NoAuthScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.NoAuthScheme.deserializeBinaryFromReader);
      msg.setNoAuth(value);
      break;
    case 2:
      var value = new proto.gendocu.sdk_generator.v3.TokenScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.TokenScheme.deserializeBinaryFromReader);
      msg.setTokenScheme(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getNoAuth();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.NoAuthScheme.serializeBinaryToWriter
    );
  }
  f = message.getTokenScheme();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.gendocu.sdk_generator.v3.TokenScheme.serializeBinaryToWriter
    );
  }
};


/**
 * optional NoAuthScheme no_auth = 1;
 * @return {?proto.gendocu.sdk_generator.v3.NoAuthScheme}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.getNoAuth = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.NoAuthScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.NoAuthScheme, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.NoAuthScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} returns this
*/
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.setNoAuth = function(value) {
  return jspb.Message.setOneofWrapperField(this, 1, proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.oneofGroups_[0], value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.clearNoAuth = function() {
  return this.setNoAuth(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.hasNoAuth = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional TokenScheme token_scheme = 2;
 * @return {?proto.gendocu.sdk_generator.v3.TokenScheme}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.getTokenScheme = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.TokenScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.TokenScheme, 2));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.TokenScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} returns this
*/
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.setTokenScheme = function(value) {
  return jspb.Message.setOneofWrapperField(this, 2, proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.oneofGroups_[0], value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.CallAuthenticationMethod} returns this
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.clearTokenScheme = function() {
  return this.setTokenScheme(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.CallAuthenticationMethod.prototype.hasTokenScheme = function() {
  return jspb.Message.getField(this, 2) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.TokenScheme.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.toObject = function(includeInstance, msg) {
  var f, obj = {
    exampleValue: jspb.Message.getFieldWithDefault(msg, 4, ""),
    grpc: (f = msg.getGrpc()) && proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.toObject(includeInstance, f),
    grpcweb: (f = msg.getGrpcweb()) && proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.toObject(includeInstance, f),
    rest: (f = msg.getRest()) && proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.TokenScheme;
  return proto.gendocu.sdk_generator.v3.TokenScheme.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setExampleValue(value);
      break;
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.deserializeBinaryFromReader);
      msg.setGrpc(value);
      break;
    case 2:
      var value = new proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.deserializeBinaryFromReader);
      msg.setGrpcweb(value);
      break;
    case 3:
      var value = new proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.deserializeBinaryFromReader);
      msg.setRest(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.TokenScheme.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getExampleValue();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getGrpc();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.serializeBinaryToWriter
    );
  }
  f = message.getGrpcweb();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.serializeBinaryToWriter
    );
  }
  f = message.getRest();
  if (f != null) {
    writer.writeMessage(
      3,
      f,
      proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.serializeBinaryToWriter
    );
  }
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.toObject = function(includeInstance, msg) {
  var f, obj = {
    meta: jspb.Message.getFieldWithDefault(msg, 1, ""),
    valuePrefix: jspb.Message.getFieldWithDefault(msg, 3, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme;
  return proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setMeta(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setValuePrefix(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMeta();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getValuePrefix();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
};


/**
 * optional string meta = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.getMeta = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.setMeta = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string value_prefix = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.getValuePrefix = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme.prototype.setValuePrefix = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.toObject = function(includeInstance, msg) {
  var f, obj = {
    header: jspb.Message.getFieldWithDefault(msg, 1, ""),
    valuePrefix: jspb.Message.getFieldWithDefault(msg, 3, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme;
  return proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setHeader(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setValuePrefix(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getHeader();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getValuePrefix();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
};


/**
 * optional string header = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.getHeader = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.setHeader = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string value_prefix = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.getValuePrefix = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme.prototype.setValuePrefix = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.toObject = function(includeInstance, msg) {
  var f, obj = {
    header: jspb.Message.getFieldWithDefault(msg, 1, ""),
    valuePrefix: jspb.Message.getFieldWithDefault(msg, 3, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme;
  return proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setHeader(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setValuePrefix(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getHeader();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getValuePrefix();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
};


/**
 * optional string header = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.getHeader = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.setHeader = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string value_prefix = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.getValuePrefix = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme.prototype.setValuePrefix = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string example_value = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.getExampleValue = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.setExampleValue = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional gRPCScheme grpc = 1;
 * @return {?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.getGrpc = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
*/
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.setGrpc = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.clearGrpc = function() {
  return this.setGrpc(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.hasGrpc = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional gRPCWebScheme grpcweb = 2;
 * @return {?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.getGrpcweb = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme, 2));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.TokenScheme.gRPCWebScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
*/
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.setGrpcweb = function(value) {
  return jspb.Message.setWrapperField(this, 2, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.clearGrpcweb = function() {
  return this.setGrpcweb(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.hasGrpcweb = function() {
  return jspb.Message.getField(this, 2) != null;
};


/**
 * optional RESTScheme rest = 3;
 * @return {?proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.getRest = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme, 3));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.TokenScheme.RESTScheme|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
*/
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.setRest = function(value) {
  return jspb.Message.setWrapperField(this, 3, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.TokenScheme} returns this
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.clearRest = function() {
  return this.setRest(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TokenScheme.prototype.hasRest = function() {
  return jspb.Message.getField(this, 3) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.NoAuthScheme.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.NoAuthScheme} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.toObject = function(includeInstance, msg) {
  var f, obj = {
    noAuth: jspb.Message.getBooleanFieldWithDefault(msg, 1, false)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.NoAuthScheme}
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.NoAuthScheme;
  return proto.gendocu.sdk_generator.v3.NoAuthScheme.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.NoAuthScheme} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.NoAuthScheme}
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setNoAuth(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.NoAuthScheme.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.NoAuthScheme} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getNoAuth();
  if (f) {
    writer.writeBool(
      1,
      f
    );
  }
};


/**
 * optional bool no_auth = 1;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.prototype.getNoAuth = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 1, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.NoAuthScheme} returns this
 */
proto.gendocu.sdk_generator.v3.NoAuthScheme.prototype.setNoAuth = function(value) {
  return jspb.Message.setProto3BooleanField(this, 1, value);
};


goog.object.extend(exports, proto.gendocu.sdk_generator.v3);
