// GendocuPublicApis/sdk/ts/gendocu/sdk_generator/v3/sdk_generator_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/sdk_generator/v3/sdk_generator.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

var gendocu_common_types_pb = require('../../../gendocu/common/types_pb.js');
goog.object.extend(proto, gendocu_common_types_pb);
var google_protobuf_descriptor_pb = require('google-protobuf/google/protobuf/descriptor_pb.js');
goog.object.extend(proto, google_protobuf_descriptor_pb);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.EnumDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.Example', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.FieldDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.IncludeInfo', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.MethodDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.ServiceDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.TypeDescription', null, global);
goog.exportSymbol('proto.gendocu.sdk_generator.v3.Version', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.Version = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.Version, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.Version.displayName = 'proto.gendocu.sdk_generator.v3.Version';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.displayName = 'proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.MethodDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.MethodDescription.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.MethodDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.MethodDescription.displayName = 'proto.gendocu.sdk_generator.v3.MethodDescription';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.IncludeInfo = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.IncludeInfo, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.IncludeInfo.displayName = 'proto.gendocu.sdk_generator.v3.IncludeInfo';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.ServiceDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.ServiceDescription.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.ServiceDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.ServiceDescription.displayName = 'proto.gendocu.sdk_generator.v3.ServiceDescription';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.TypeDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.sdk_generator.v3.TypeDescription.repeatedFields_, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.TypeDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.TypeDescription.displayName = 'proto.gendocu.sdk_generator.v3.TypeDescription';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.Example = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.Example, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.Example.displayName = 'proto.gendocu.sdk_generator.v3.Example';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.EnumDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.EnumDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.EnumDescription.displayName = 'proto.gendocu.sdk_generator.v3.EnumDescription';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.sdk_generator.v3.FieldDescription = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.sdk_generator.v3.FieldDescription, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.sdk_generator.v3.FieldDescription.displayName = 'proto.gendocu.sdk_generator.v3.FieldDescription';
}



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.Version.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.Version.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.Version} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Version.toObject = function(includeInstance, msg) {
  var f, obj = {
    major: jspb.Message.getFieldWithDefault(msg, 1, 0),
    minor: jspb.Message.getFieldWithDefault(msg, 2, 0),
    patch: jspb.Message.getFieldWithDefault(msg, 3, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.Version}
 */
proto.gendocu.sdk_generator.v3.Version.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.Version;
  return proto.gendocu.sdk_generator.v3.Version.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.Version} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.Version}
 */
proto.gendocu.sdk_generator.v3.Version.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {number} */ (reader.readInt32());
      msg.setMajor(value);
      break;
    case 2:
      var value = /** @type {number} */ (reader.readInt32());
      msg.setMinor(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setPatch(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.Version.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.Version.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.Version} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Version.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMajor();
  if (f !== 0) {
    writer.writeInt32(
      1,
      f
    );
  }
  f = message.getMinor();
  if (f !== 0) {
    writer.writeInt32(
      2,
      f
    );
  }
  f = message.getPatch();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
};


/**
 * optional int32 major = 1;
 * @return {number}
 */
proto.gendocu.sdk_generator.v3.Version.prototype.getMajor = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0));
};


/**
 * @param {number} value
 * @return {!proto.gendocu.sdk_generator.v3.Version} returns this
 */
proto.gendocu.sdk_generator.v3.Version.prototype.setMajor = function(value) {
  return jspb.Message.setProto3IntField(this, 1, value);
};


/**
 * optional int32 minor = 2;
 * @return {number}
 */
proto.gendocu.sdk_generator.v3.Version.prototype.getMinor = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
};


/**
 * @param {number} value
 * @return {!proto.gendocu.sdk_generator.v3.Version} returns this
 */
proto.gendocu.sdk_generator.v3.Version.prototype.setMinor = function(value) {
  return jspb.Message.setProto3IntField(this, 2, value);
};


/**
 * optional string patch = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Version.prototype.getPatch = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Version} returns this
 */
proto.gendocu.sdk_generator.v3.Version.prototype.setPatch = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.toObject = function(includeInstance, msg) {
  var f, obj = {
    version: (f = msg.getVersion()) && proto.gendocu.sdk_generator.v3.Version.toObject(includeInstance, f),
    description: jspb.Message.getFieldWithDefault(msg, 2, ""),
    authenticationDescription: jspb.Message.getFieldWithDefault(msg, 6, ""),
    methodsMap: (f = msg.getMethodsMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.MethodDescription.toObject) : [],
    servicesMap: (f = msg.getServicesMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.ServiceDescription.toObject) : [],
    typesMap: (f = msg.getTypesMap()) ? f.toObject(includeInstance, proto.gendocu.sdk_generator.v3.TypeDescription.toObject) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel;
  return proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = new proto.gendocu.sdk_generator.v3.Version;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.Version.deserializeBinaryFromReader);
      msg.setVersion(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setAuthenticationDescription(value);
      break;
    case 3:
      var value = msg.getMethodsMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.MethodDescription.deserializeBinaryFromReader, "", new proto.gendocu.sdk_generator.v3.MethodDescription());
         });
      break;
    case 4:
      var value = msg.getServicesMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.ServiceDescription.deserializeBinaryFromReader, "", new proto.gendocu.sdk_generator.v3.ServiceDescription());
         });
      break;
    case 5:
      var value = msg.getTypesMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readMessage, proto.gendocu.sdk_generator.v3.TypeDescription.deserializeBinaryFromReader, "", new proto.gendocu.sdk_generator.v3.TypeDescription());
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getVersion();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      proto.gendocu.sdk_generator.v3.Version.serializeBinaryToWriter
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getAuthenticationDescription();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getMethodsMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(3, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.MethodDescription.serializeBinaryToWriter);
  }
  f = message.getServicesMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(4, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.ServiceDescription.serializeBinaryToWriter);
  }
  f = message.getTypesMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(5, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeMessage, proto.gendocu.sdk_generator.v3.TypeDescription.serializeBinaryToWriter);
  }
};


/**
 * optional Version version = 1;
 * @return {?proto.gendocu.sdk_generator.v3.Version}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getVersion = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.Version} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.Version, 1));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.Version|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
*/
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.setVersion = function(value) {
  return jspb.Message.setWrapperField(this, 1, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.clearVersion = function() {
  return this.setVersion(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.hasVersion = function() {
  return jspb.Message.getField(this, 1) != null;
};


/**
 * optional string description = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string authentication_description = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getAuthenticationDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.setAuthenticationDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * map<string, MethodDescription> methods = 3;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.MethodDescription>}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getMethodsMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.MethodDescription>} */ (
      jspb.Message.getMapField(this, 3, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.MethodDescription));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.clearMethodsMap = function() {
  this.getMethodsMap().clear();
  return this;};


/**
 * map<string, ServiceDescription> services = 4;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.ServiceDescription>}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getServicesMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.ServiceDescription>} */ (
      jspb.Message.getMapField(this, 4, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.ServiceDescription));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.clearServicesMap = function() {
  this.getServicesMap().clear();
  return this;};


/**
 * map<string, TypeDescription> types = 5;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.TypeDescription>}
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.getTypesMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,!proto.gendocu.sdk_generator.v3.TypeDescription>} */ (
      jspb.Message.getMapField(this, 5, opt_noLazyCreate,
      proto.gendocu.sdk_generator.v3.TypeDescription));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel} returns this
 */
proto.gendocu.sdk_generator.v3.APIDescriptionGeneratorModel.prototype.clearTypesMap = function() {
  this.getTypesMap().clear();
  return this;};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.MethodDescription.repeatedFields_ = [8];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.MethodDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.MethodDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.MethodDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    methodId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    serviceId: jspb.Message.getFieldWithDefault(msg, 2, ""),
    name: jspb.Message.getFieldWithDefault(msg, 3, ""),
    description: jspb.Message.getFieldWithDefault(msg, 4, ""),
    sourceProtoFile: jspb.Message.getFieldWithDefault(msg, 5, ""),
    inTypeId: jspb.Message.getFieldWithDefault(msg, 6, ""),
    outTypeId: jspb.Message.getFieldWithDefault(msg, 7, ""),
    includeInfoList: jspb.Message.toObjectList(msg.getIncludeInfoList(),
    proto.gendocu.sdk_generator.v3.IncludeInfo.toObject, includeInstance),
    inStreaming: jspb.Message.getBooleanFieldWithDefault(msg, 9, false),
    outStreaming: jspb.Message.getBooleanFieldWithDefault(msg, 10, false)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.MethodDescription;
  return proto.gendocu.sdk_generator.v3.MethodDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.MethodDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setMethodId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setServiceId(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setName(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setSourceProtoFile(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setInTypeId(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setOutTypeId(value);
      break;
    case 8:
      var value = new proto.gendocu.sdk_generator.v3.IncludeInfo;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinaryFromReader);
      msg.addIncludeInfo(value);
      break;
    case 9:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setInStreaming(value);
      break;
    case 10:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setOutStreaming(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.MethodDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.MethodDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.MethodDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getMethodId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getServiceId();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getName();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getSourceProtoFile();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
  f = message.getInTypeId();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getOutTypeId();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getIncludeInfoList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      8,
      f,
      proto.gendocu.sdk_generator.v3.IncludeInfo.serializeBinaryToWriter
    );
  }
  f = message.getInStreaming();
  if (f) {
    writer.writeBool(
      9,
      f
    );
  }
  f = message.getOutStreaming();
  if (f) {
    writer.writeBool(
      10,
      f
    );
  }
};


/**
 * optional string method_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getMethodId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setMethodId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string service_id = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getServiceId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setServiceId = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string name = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setName = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string description = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional string source_proto_file = 5;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getSourceProtoFile = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setSourceProtoFile = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};


/**
 * optional string in_type_id = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getInTypeId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setInTypeId = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string out_type_id = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getOutTypeId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setOutTypeId = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * repeated IncludeInfo include_info = 8;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getIncludeInfoList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.IncludeInfo, 8));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
*/
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setIncludeInfoList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 8, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.addIncludeInfo = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 8, opt_value, proto.gendocu.sdk_generator.v3.IncludeInfo, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.clearIncludeInfoList = function() {
  return this.setIncludeInfoList([]);
};


/**
 * optional bool in_streaming = 9;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getInStreaming = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 9, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setInStreaming = function(value) {
  return jspb.Message.setProto3BooleanField(this, 9, value);
};


/**
 * optional bool out_streaming = 10;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.getOutStreaming = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 10, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.MethodDescription} returns this
 */
proto.gendocu.sdk_generator.v3.MethodDescription.prototype.setOutStreaming = function(value) {
  return jspb.Message.setProto3BooleanField(this, 10, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.IncludeInfo.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.toObject = function(includeInstance, msg) {
  var f, obj = {
    sdkProvider: jspb.Message.getFieldWithDefault(msg, 1, 0),
    protoPkg: jspb.Message.getFieldWithDefault(msg, 2, ""),
    goImportPath: jspb.Message.getFieldWithDefault(msg, 3, ""),
    goPkg: jspb.Message.getFieldWithDefault(msg, 4, ""),
    tsImportPath: jspb.Message.getFieldWithDefault(msg, 5, ""),
    javaServicePkg: jspb.Message.getFieldWithDefault(msg, 6, ""),
    javaMessagePkg: jspb.Message.getFieldWithDefault(msg, 8, ""),
    csharpPkg: jspb.Message.getFieldWithDefault(msg, 7, ""),
    rustCrate: jspb.Message.getFieldWithDefault(msg, 9, ""),
    rustPackage: jspb.Message.getFieldWithDefault(msg, 10, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.IncludeInfo;
  return proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {!proto.gendocu.common.SDKProvider} */ (reader.readEnum());
      msg.setSdkProvider(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setProtoPkg(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setGoImportPath(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setGoPkg(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setTsImportPath(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setJavaServicePkg(value);
      break;
    case 8:
      var value = /** @type {string} */ (reader.readString());
      msg.setJavaMessagePkg(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setCsharpPkg(value);
      break;
    case 9:
      var value = /** @type {string} */ (reader.readString());
      msg.setRustCrate(value);
      break;
    case 10:
      var value = /** @type {string} */ (reader.readString());
      msg.setRustPackage(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.IncludeInfo.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getSdkProvider();
  if (f !== 0.0) {
    writer.writeEnum(
      1,
      f
    );
  }
  f = message.getProtoPkg();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getGoImportPath();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getGoPkg();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getTsImportPath();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
  f = message.getJavaServicePkg();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getJavaMessagePkg();
  if (f.length > 0) {
    writer.writeString(
      8,
      f
    );
  }
  f = message.getCsharpPkg();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getRustCrate();
  if (f.length > 0) {
    writer.writeString(
      9,
      f
    );
  }
  f = message.getRustPackage();
  if (f.length > 0) {
    writer.writeString(
      10,
      f
    );
  }
};


/**
 * optional gendocu.common.SDKProvider sdk_provider = 1;
 * @return {!proto.gendocu.common.SDKProvider}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getSdkProvider = function() {
  return /** @type {!proto.gendocu.common.SDKProvider} */ (jspb.Message.getFieldWithDefault(this, 1, 0));
};


/**
 * @param {!proto.gendocu.common.SDKProvider} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setSdkProvider = function(value) {
  return jspb.Message.setProto3EnumField(this, 1, value);
};


/**
 * optional string proto_pkg = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getProtoPkg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setProtoPkg = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string go_import_path = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getGoImportPath = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setGoImportPath = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string go_pkg = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getGoPkg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setGoPkg = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional string ts_import_path = 5;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getTsImportPath = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setTsImportPath = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};


/**
 * optional string java_service_pkg = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getJavaServicePkg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setJavaServicePkg = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string java_message_pkg = 8;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getJavaMessagePkg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 8, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setJavaMessagePkg = function(value) {
  return jspb.Message.setProto3StringField(this, 8, value);
};


/**
 * optional string csharp_pkg = 7;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getCsharpPkg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setCsharpPkg = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional string rust_crate = 9;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getRustCrate = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 9, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setRustCrate = function(value) {
  return jspb.Message.setProto3StringField(this, 9, value);
};


/**
 * optional string rust_package = 10;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.getRustPackage = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 10, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo} returns this
 */
proto.gendocu.sdk_generator.v3.IncludeInfo.prototype.setRustPackage = function(value) {
  return jspb.Message.setProto3StringField(this, 10, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.repeatedFields_ = [5];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.ServiceDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.ServiceDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    serviceId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    shortName: jspb.Message.getFieldWithDefault(msg, 2, ""),
    fullName: jspb.Message.getFieldWithDefault(msg, 6, ""),
    description: jspb.Message.getFieldWithDefault(msg, 3, ""),
    sourceProtoFile: jspb.Message.getFieldWithDefault(msg, 4, ""),
    includeInfoList: jspb.Message.toObjectList(msg.getIncludeInfoList(),
    proto.gendocu.sdk_generator.v3.IncludeInfo.toObject, includeInstance)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.ServiceDescription;
  return proto.gendocu.sdk_generator.v3.ServiceDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.ServiceDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setServiceId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setShortName(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setFullName(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setSourceProtoFile(value);
      break;
    case 5:
      var value = new proto.gendocu.sdk_generator.v3.IncludeInfo;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinaryFromReader);
      msg.addIncludeInfo(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.ServiceDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.ServiceDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getServiceId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getShortName();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getFullName();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getSourceProtoFile();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getIncludeInfoList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      5,
      f,
      proto.gendocu.sdk_generator.v3.IncludeInfo.serializeBinaryToWriter
    );
  }
};


/**
 * optional string service_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getServiceId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setServiceId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string short_name = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getShortName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setShortName = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string full_name = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getFullName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setFullName = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string description = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string source_proto_file = 4;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getSourceProtoFile = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setSourceProtoFile = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * repeated IncludeInfo include_info = 5;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.getIncludeInfoList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.IncludeInfo, 5));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} value
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
*/
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.setIncludeInfoList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 5, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo}
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.addIncludeInfo = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 5, opt_value, proto.gendocu.sdk_generator.v3.IncludeInfo, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.ServiceDescription} returns this
 */
proto.gendocu.sdk_generator.v3.ServiceDescription.prototype.clearIncludeInfoList = function() {
  return this.setIncludeInfoList([]);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.sdk_generator.v3.TypeDescription.repeatedFields_ = [4,5,12];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.TypeDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.TypeDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TypeDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    typeId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    shortName: jspb.Message.getFieldWithDefault(msg, 2, ""),
    fullName: jspb.Message.getFieldWithDefault(msg, 9, ""),
    sourceProtoFile: jspb.Message.getFieldWithDefault(msg, 3, ""),
    includeInfoList: jspb.Message.toObjectList(msg.getIncludeInfoList(),
    proto.gendocu.sdk_generator.v3.IncludeInfo.toObject, includeInstance),
    fieldsList: jspb.Message.toObjectList(msg.getFieldsList(),
    proto.gendocu.sdk_generator.v3.FieldDescription.toObject, includeInstance),
    description: jspb.Message.getFieldWithDefault(msg, 6, ""),
    isMapType: jspb.Message.getBooleanFieldWithDefault(msg, 7, false),
    enumDescription: (f = msg.getEnumDescription()) && proto.gendocu.sdk_generator.v3.EnumDescription.toObject(includeInstance, f),
    pbDescriptor: (f = msg.getPbDescriptor()) && google_protobuf_descriptor_pb.DescriptorProto.toObject(includeInstance, f),
    subtypeOf: jspb.Message.getFieldWithDefault(msg, 11, ""),
    examplesList: jspb.Message.toObjectList(msg.getExamplesList(),
    proto.gendocu.sdk_generator.v3.Example.toObject, includeInstance)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.TypeDescription;
  return proto.gendocu.sdk_generator.v3.TypeDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.TypeDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setTypeId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setShortName(value);
      break;
    case 9:
      var value = /** @type {string} */ (reader.readString());
      msg.setFullName(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setSourceProtoFile(value);
      break;
    case 4:
      var value = new proto.gendocu.sdk_generator.v3.IncludeInfo;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.IncludeInfo.deserializeBinaryFromReader);
      msg.addIncludeInfo(value);
      break;
    case 5:
      var value = new proto.gendocu.sdk_generator.v3.FieldDescription;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.FieldDescription.deserializeBinaryFromReader);
      msg.addFields(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 7:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setIsMapType(value);
      break;
    case 8:
      var value = new proto.gendocu.sdk_generator.v3.EnumDescription;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.EnumDescription.deserializeBinaryFromReader);
      msg.setEnumDescription(value);
      break;
    case 10:
      var value = new google_protobuf_descriptor_pb.DescriptorProto;
      reader.readMessage(value,google_protobuf_descriptor_pb.DescriptorProto.deserializeBinaryFromReader);
      msg.setPbDescriptor(value);
      break;
    case 11:
      var value = /** @type {string} */ (reader.readString());
      msg.setSubtypeOf(value);
      break;
    case 12:
      var value = new proto.gendocu.sdk_generator.v3.Example;
      reader.readMessage(value,proto.gendocu.sdk_generator.v3.Example.deserializeBinaryFromReader);
      msg.addExamples(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.TypeDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.TypeDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.TypeDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getTypeId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getShortName();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getFullName();
  if (f.length > 0) {
    writer.writeString(
      9,
      f
    );
  }
  f = message.getSourceProtoFile();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getIncludeInfoList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      4,
      f,
      proto.gendocu.sdk_generator.v3.IncludeInfo.serializeBinaryToWriter
    );
  }
  f = message.getFieldsList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      5,
      f,
      proto.gendocu.sdk_generator.v3.FieldDescription.serializeBinaryToWriter
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getIsMapType();
  if (f) {
    writer.writeBool(
      7,
      f
    );
  }
  f = message.getEnumDescription();
  if (f != null) {
    writer.writeMessage(
      8,
      f,
      proto.gendocu.sdk_generator.v3.EnumDescription.serializeBinaryToWriter
    );
  }
  f = message.getPbDescriptor();
  if (f != null) {
    writer.writeMessage(
      10,
      f,
      google_protobuf_descriptor_pb.DescriptorProto.serializeBinaryToWriter
    );
  }
  f = message.getSubtypeOf();
  if (f.length > 0) {
    writer.writeString(
      11,
      f
    );
  }
  f = message.getExamplesList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      12,
      f,
      proto.gendocu.sdk_generator.v3.Example.serializeBinaryToWriter
    );
  }
};


/**
 * optional string type_id = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getTypeId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setTypeId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string short_name = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getShortName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setShortName = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string full_name = 9;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getFullName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 9, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setFullName = function(value) {
  return jspb.Message.setProto3StringField(this, 9, value);
};


/**
 * optional string source_proto_file = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getSourceProtoFile = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setSourceProtoFile = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * repeated IncludeInfo include_info = 4;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getIncludeInfoList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.IncludeInfo, 4));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.IncludeInfo>} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
*/
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setIncludeInfoList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 4, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.IncludeInfo=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.IncludeInfo}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.addIncludeInfo = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 4, opt_value, proto.gendocu.sdk_generator.v3.IncludeInfo, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.clearIncludeInfoList = function() {
  return this.setIncludeInfoList([]);
};


/**
 * repeated FieldDescription fields = 5;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.FieldDescription>}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getFieldsList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.FieldDescription>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.FieldDescription, 5));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.FieldDescription>} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
*/
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setFieldsList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 5, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.FieldDescription=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.addFields = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 5, opt_value, proto.gendocu.sdk_generator.v3.FieldDescription, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.clearFieldsList = function() {
  return this.setFieldsList([]);
};


/**
 * optional string description = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional bool is_map_type = 7;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getIsMapType = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 7, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setIsMapType = function(value) {
  return jspb.Message.setProto3BooleanField(this, 7, value);
};


/**
 * optional EnumDescription enum_description = 8;
 * @return {?proto.gendocu.sdk_generator.v3.EnumDescription}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getEnumDescription = function() {
  return /** @type{?proto.gendocu.sdk_generator.v3.EnumDescription} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.sdk_generator.v3.EnumDescription, 8));
};


/**
 * @param {?proto.gendocu.sdk_generator.v3.EnumDescription|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
*/
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setEnumDescription = function(value) {
  return jspb.Message.setWrapperField(this, 8, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.clearEnumDescription = function() {
  return this.setEnumDescription(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.hasEnumDescription = function() {
  return jspb.Message.getField(this, 8) != null;
};


/**
 * optional google.protobuf.DescriptorProto pb_descriptor = 10;
 * @return {?proto.google.protobuf.DescriptorProto}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getPbDescriptor = function() {
  return /** @type{?proto.google.protobuf.DescriptorProto} */ (
    jspb.Message.getWrapperField(this, google_protobuf_descriptor_pb.DescriptorProto, 10));
};


/**
 * @param {?proto.google.protobuf.DescriptorProto|undefined} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
*/
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setPbDescriptor = function(value) {
  return jspb.Message.setWrapperField(this, 10, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.clearPbDescriptor = function() {
  return this.setPbDescriptor(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.hasPbDescriptor = function() {
  return jspb.Message.getField(this, 10) != null;
};


/**
 * optional string subtype_of = 11;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getSubtypeOf = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 11, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setSubtypeOf = function(value) {
  return jspb.Message.setProto3StringField(this, 11, value);
};


/**
 * repeated Example examples = 12;
 * @return {!Array<!proto.gendocu.sdk_generator.v3.Example>}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.getExamplesList = function() {
  return /** @type{!Array<!proto.gendocu.sdk_generator.v3.Example>} */ (
    jspb.Message.getRepeatedWrapperField(this, proto.gendocu.sdk_generator.v3.Example, 12));
};


/**
 * @param {!Array<!proto.gendocu.sdk_generator.v3.Example>} value
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
*/
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.setExamplesList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 12, value);
};


/**
 * @param {!proto.gendocu.sdk_generator.v3.Example=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.sdk_generator.v3.Example}
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.addExamples = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 12, opt_value, proto.gendocu.sdk_generator.v3.Example, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.sdk_generator.v3.TypeDescription} returns this
 */
proto.gendocu.sdk_generator.v3.TypeDescription.prototype.clearExamplesList = function() {
  return this.setExamplesList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.Example.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.Example.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.Example} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Example.toObject = function(includeInstance, msg) {
  var f, obj = {
    pb_default: jspb.Message.getBooleanFieldWithDefault(msg, 1, false),
    encoded: jspb.Message.getFieldWithDefault(msg, 2, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.Example}
 */
proto.gendocu.sdk_generator.v3.Example.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.Example;
  return proto.gendocu.sdk_generator.v3.Example.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.Example} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.Example}
 */
proto.gendocu.sdk_generator.v3.Example.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setDefault(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setEncoded(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.Example.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.Example.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.Example} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.Example.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getDefault();
  if (f) {
    writer.writeBool(
      1,
      f
    );
  }
  f = message.getEncoded();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
};


/**
 * optional bool default = 1;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.Example.prototype.getDefault = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 1, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.Example} returns this
 */
proto.gendocu.sdk_generator.v3.Example.prototype.setDefault = function(value) {
  return jspb.Message.setProto3BooleanField(this, 1, value);
};


/**
 * optional string encoded = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.Example.prototype.getEncoded = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.Example} returns this
 */
proto.gendocu.sdk_generator.v3.Example.prototype.setEncoded = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.EnumDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.EnumDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnumDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    valuesMap: (f = msg.getValuesMap()) ? f.toObject(includeInstance, undefined) : [],
    descriptionMap: (f = msg.getDescriptionMap()) ? f.toObject(includeInstance, undefined) : []
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.EnumDescription}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.EnumDescription;
  return proto.gendocu.sdk_generator.v3.EnumDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.EnumDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.EnumDescription}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = msg.getValuesMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readInt32, null, "", 0);
         });
      break;
    case 2:
      var value = msg.getDescriptionMap();
      reader.readMessage(value, function(message, reader) {
        jspb.Map.deserializeBinary(message, reader, jspb.BinaryReader.prototype.readString, jspb.BinaryReader.prototype.readString, null, "", "");
         });
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.EnumDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.EnumDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.EnumDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getValuesMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(1, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeInt32);
  }
  f = message.getDescriptionMap(true);
  if (f && f.getLength() > 0) {
    f.serializeBinary(2, writer, jspb.BinaryWriter.prototype.writeString, jspb.BinaryWriter.prototype.writeString);
  }
};


/**
 * map<string, int32> values = 1;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,number>}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.getValuesMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,number>} */ (
      jspb.Message.getMapField(this, 1, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.EnumDescription} returns this
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.clearValuesMap = function() {
  this.getValuesMap().clear();
  return this;};


/**
 * map<string, string> description = 2;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,string>}
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.getDescriptionMap = function(opt_noLazyCreate) {
  return /** @type {!jspb.Map<string,string>} */ (
      jspb.Message.getMapField(this, 2, opt_noLazyCreate,
      null));
};


/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.gendocu.sdk_generator.v3.EnumDescription} returns this
 */
proto.gendocu.sdk_generator.v3.EnumDescription.prototype.clearDescriptionMap = function() {
  this.getDescriptionMap().clear();
  return this;};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.sdk_generator.v3.FieldDescription.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.sdk_generator.v3.FieldDescription} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.FieldDescription.toObject = function(includeInstance, msg) {
  var f, obj = {
    name: jspb.Message.getFieldWithDefault(msg, 1, ""),
    typeId: jspb.Message.getFieldWithDefault(msg, 2, ""),
    description: jspb.Message.getFieldWithDefault(msg, 3, ""),
    repeated: jspb.Message.getBooleanFieldWithDefault(msg, 4, false),
    required: jspb.Message.getBooleanFieldWithDefault(msg, 5, false),
    optional: jspb.Message.getBooleanFieldWithDefault(msg, 7, false),
    proto3Optional: jspb.Message.getBooleanFieldWithDefault(msg, 8, false),
    oneOf: jspb.Message.getFieldWithDefault(msg, 6, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.sdk_generator.v3.FieldDescription;
  return proto.gendocu.sdk_generator.v3.FieldDescription.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.sdk_generator.v3.FieldDescription} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setName(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setTypeId(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setDescription(value);
      break;
    case 4:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setRepeated(value);
      break;
    case 5:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setRequired(value);
      break;
    case 7:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setOptional(value);
      break;
    case 8:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setProto3Optional(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setOneOf(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.sdk_generator.v3.FieldDescription.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.sdk_generator.v3.FieldDescription} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.sdk_generator.v3.FieldDescription.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getName();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getTypeId();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getDescription();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getRepeated();
  if (f) {
    writer.writeBool(
      4,
      f
    );
  }
  f = message.getRequired();
  if (f) {
    writer.writeBool(
      5,
      f
    );
  }
  f = message.getOptional();
  if (f) {
    writer.writeBool(
      7,
      f
    );
  }
  f = message.getProto3Optional();
  if (f) {
    writer.writeBool(
      8,
      f
    );
  }
  f = message.getOneOf();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
};


/**
 * optional string name = 1;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setName = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string type_id = 2;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getTypeId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setTypeId = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional string description = 3;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getDescription = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setDescription = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional bool repeated = 4;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getRepeated = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 4, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setRepeated = function(value) {
  return jspb.Message.setProto3BooleanField(this, 4, value);
};


/**
 * optional bool required = 5;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getRequired = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 5, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setRequired = function(value) {
  return jspb.Message.setProto3BooleanField(this, 5, value);
};


/**
 * optional bool optional = 7;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getOptional = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 7, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setOptional = function(value) {
  return jspb.Message.setProto3BooleanField(this, 7, value);
};


/**
 * optional bool proto3_optional = 8;
 * @return {boolean}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getProto3Optional = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 8, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setProto3Optional = function(value) {
  return jspb.Message.setProto3BooleanField(this, 8, value);
};


/**
 * optional string one_of = 6;
 * @return {string}
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.getOneOf = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.sdk_generator.v3.FieldDescription} returns this
 */
proto.gendocu.sdk_generator.v3.FieldDescription.prototype.setOneOf = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


goog.object.extend(exports, proto.gendocu.sdk_generator.v3);
