// @useoptic/optic/build/commands/oas/operations/diffs/traversers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpecOperationDiffTraverser = void 0;
var _queries = require("../../../capture/operations/queries");
var _index = require("./visitors/index");
var _tsResults = require("ts-results");
class SpecOperationDiffTraverser {
  traverse(operation, spec) {
    this.operation = operation;
    this.spec = spec;
    this.queries = new _queries.OperationQueries((0, _queries.specToOperations)(spec));
  }
  *results() {
    if (!this.operation || !this.queries || !this.spec) return;
    const {
      operation,
      queries,
      spec
    } = this;
    let pathPatternOption = queries.findPathPattern(operation.pathPattern).expect('path pattern should be able to be matched against spec');
    let pathSpecOption = pathPatternOption.map(pathPattern => spec.paths[pathPattern]);
    yield* (0, _index.visitPath)(operation.pathPattern, pathSpecOption, {
      pathPattern: pathPatternOption
    });
    if (pathSpecOption.some && pathPatternOption.some) {
      let pathSpec = pathSpecOption.val;
      let pathPattern = pathPatternOption.val;
      for (let method of operation.methods) {
        let methodSpec = pathSpec[method];
        yield* (0, _index.visitMethod)(method, methodSpec ? (0, _tsResults.Some)(methodSpec) : _tsResults.None, {
          pathPattern
        });
      }
    }
  }
}
exports.SpecOperationDiffTraverser = SpecOperationDiffTraverser;
