// GendocuPublicApis/sdk/ts/gendocu/documentation/v1/sections_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/documentation/v1/sections.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

goog.exportSymbol('proto.gendocu.documentation.v1.CustomSection', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.CustomSection.Position', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.CustomSection.Type', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.CustomSection = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.documentation.v1.CustomSection, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.CustomSection.displayName = 'proto.gendocu.documentation.v1.CustomSection';
}



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.CustomSection.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.CustomSection} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.CustomSection.toObject = function(includeInstance, msg) {
  var f, obj = {
    id: jspb.Message.getFieldWithDefault(msg, 1, ""),
    order: jspb.Message.getFieldWithDefault(msg, 2, 0),
    position: jspb.Message.getFieldWithDefault(msg, 3, 0),
    leftTitle: jspb.Message.getFieldWithDefault(msg, 4, ""),
    leftContent: jspb.Message.getFieldWithDefault(msg, 5, ""),
    leftContentHtml: jspb.Message.getFieldWithDefault(msg, 100, ""),
    hasIsland: jspb.Message.getBooleanFieldWithDefault(msg, 6, false),
    islandTitle: jspb.Message.getFieldWithDefault(msg, 7, ""),
    islandContent: jspb.Message.getFieldWithDefault(msg, 8, ""),
    islandContentHtml: jspb.Message.getFieldWithDefault(msg, 101, ""),
    type: jspb.Message.getFieldWithDefault(msg, 9, 0),
    hidden: jspb.Message.getBooleanFieldWithDefault(msg, 11, false)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.CustomSection}
 */
proto.gendocu.documentation.v1.CustomSection.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.CustomSection;
  return proto.gendocu.documentation.v1.CustomSection.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.CustomSection} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.CustomSection}
 */
proto.gendocu.documentation.v1.CustomSection.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setId(value);
      break;
    case 2:
      var value = /** @type {number} */ (reader.readUint32());
      msg.setOrder(value);
      break;
    case 3:
      var value = /** @type {!proto.gendocu.documentation.v1.CustomSection.Position} */ (reader.readEnum());
      msg.setPosition(value);
      break;
    case 4:
      var value = /** @type {string} */ (reader.readString());
      msg.setLeftTitle(value);
      break;
    case 5:
      var value = /** @type {string} */ (reader.readString());
      msg.setLeftContent(value);
      break;
    case 100:
      var value = /** @type {string} */ (reader.readString());
      msg.setLeftContentHtml(value);
      break;
    case 6:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setHasIsland(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setIslandTitle(value);
      break;
    case 8:
      var value = /** @type {string} */ (reader.readString());
      msg.setIslandContent(value);
      break;
    case 101:
      var value = /** @type {string} */ (reader.readString());
      msg.setIslandContentHtml(value);
      break;
    case 9:
      var value = /** @type {!proto.gendocu.documentation.v1.CustomSection.Type} */ (reader.readEnum());
      msg.setType(value);
      break;
    case 11:
      var value = /** @type {boolean} */ (reader.readBool());
      msg.setHidden(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.CustomSection.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.CustomSection} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.CustomSection.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getOrder();
  if (f !== 0) {
    writer.writeUint32(
      2,
      f
    );
  }
  f = message.getPosition();
  if (f !== 0.0) {
    writer.writeEnum(
      3,
      f
    );
  }
  f = message.getLeftTitle();
  if (f.length > 0) {
    writer.writeString(
      4,
      f
    );
  }
  f = message.getLeftContent();
  if (f.length > 0) {
    writer.writeString(
      5,
      f
    );
  }
  f = message.getLeftContentHtml();
  if (f.length > 0) {
    writer.writeString(
      100,
      f
    );
  }
  f = message.getHasIsland();
  if (f) {
    writer.writeBool(
      6,
      f
    );
  }
  f = message.getIslandTitle();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getIslandContent();
  if (f.length > 0) {
    writer.writeString(
      8,
      f
    );
  }
  f = message.getIslandContentHtml();
  if (f.length > 0) {
    writer.writeString(
      101,
      f
    );
  }
  f = message.getType();
  if (f !== 0.0) {
    writer.writeEnum(
      9,
      f
    );
  }
  f = message.getHidden();
  if (f) {
    writer.writeBool(
      11,
      f
    );
  }
};


/**
 * @enum {number}
 */
proto.gendocu.documentation.v1.CustomSection.Position = {
  POSITION_UNKNOWN: 0,
  POSITION_BEFORE_OVERVIEW: 5,
  POSITION_BEFORE_CLIENT_LIBRARIES: 1,
  POSITION_BEFORE_AUTHENTICATION: 2,
  POSITION_BEFORE_API_SERVICES: 3,
  POSITION_AFTER_API_SERVICES: 4,
  CLIENT_SECTION_OVERRIDE: 10,
  AUTHENTICATION_SECTION_OVERRIDE: 11
};

/**
 * @enum {number}
 */
proto.gendocu.documentation.v1.CustomSection.Type = {
  TYPE_UNKNOWN: 0,
  TYPE_SECTION: 1,
  TYPE_SUBSECTION: 2
};

/**
 * optional string id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional uint32 order = 2;
 * @return {number}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getOrder = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
};


/**
 * @param {number} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setOrder = function(value) {
  return jspb.Message.setProto3IntField(this, 2, value);
};


/**
 * optional Position position = 3;
 * @return {!proto.gendocu.documentation.v1.CustomSection.Position}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getPosition = function() {
  return /** @type {!proto.gendocu.documentation.v1.CustomSection.Position} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};


/**
 * @param {!proto.gendocu.documentation.v1.CustomSection.Position} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setPosition = function(value) {
  return jspb.Message.setProto3EnumField(this, 3, value);
};


/**
 * optional string left_title = 4;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getLeftTitle = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setLeftTitle = function(value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};


/**
 * optional string left_content = 5;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getLeftContent = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setLeftContent = function(value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};


/**
 * optional string left_content_html = 100;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getLeftContentHtml = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 100, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setLeftContentHtml = function(value) {
  return jspb.Message.setProto3StringField(this, 100, value);
};


/**
 * optional bool has_island = 6;
 * @return {boolean}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getHasIsland = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 6, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setHasIsland = function(value) {
  return jspb.Message.setProto3BooleanField(this, 6, value);
};


/**
 * optional string island_title = 7;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getIslandTitle = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setIslandTitle = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional string island_content = 8;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getIslandContent = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 8, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setIslandContent = function(value) {
  return jspb.Message.setProto3StringField(this, 8, value);
};


/**
 * optional string island_content_html = 101;
 * @return {string}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getIslandContentHtml = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 101, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setIslandContentHtml = function(value) {
  return jspb.Message.setProto3StringField(this, 101, value);
};


/**
 * optional Type type = 9;
 * @return {!proto.gendocu.documentation.v1.CustomSection.Type}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getType = function() {
  return /** @type {!proto.gendocu.documentation.v1.CustomSection.Type} */ (jspb.Message.getFieldWithDefault(this, 9, 0));
};


/**
 * @param {!proto.gendocu.documentation.v1.CustomSection.Type} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setType = function(value) {
  return jspb.Message.setProto3EnumField(this, 9, value);
};


/**
 * optional bool hidden = 11;
 * @return {boolean}
 */
proto.gendocu.documentation.v1.CustomSection.prototype.getHidden = function() {
  return /** @type {boolean} */ (jspb.Message.getBooleanFieldWithDefault(this, 11, false));
};


/**
 * @param {boolean} value
 * @return {!proto.gendocu.documentation.v1.CustomSection} returns this
 */
proto.gendocu.documentation.v1.CustomSection.prototype.setHidden = function(value) {
  return jspb.Message.setProto3BooleanField(this, 11, value);
};


goog.object.extend(exports, proto.gendocu.documentation.v1);
