// GendocuPublicApis/sdk/ts/gendocu/documentation/v1/styling_pb.js
/* eslint-disable */
/*Generated by GenDocu.com*/
// source: gendocu/documentation/v1/styling.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

var gendocu_documentation_v1_sections_pb = require('../../../gendocu/documentation/v1/sections_pb.js');
goog.object.extend(proto, gendocu_documentation_v1_sections_pb);
goog.exportSymbol('proto.gendocu.documentation.v1.RenderMarkdownRequest', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.RenderMarkdownResponse', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.StylingSettings', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.ThemeSettings', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest', null, global);
goog.exportSymbol('proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.displayName = 'proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.displayName = 'proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.StylingSettings = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.documentation.v1.StylingSettings.repeatedFields_, null);
};
goog.inherits(proto.gendocu.documentation.v1.StylingSettings, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.StylingSettings.displayName = 'proto.gendocu.documentation.v1.StylingSettings';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.ThemeSettings = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.gendocu.documentation.v1.ThemeSettings, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.ThemeSettings.displayName = 'proto.gendocu.documentation.v1.ThemeSettings';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.documentation.v1.RenderMarkdownRequest.repeatedFields_, null);
};
goog.inherits(proto.gendocu.documentation.v1.RenderMarkdownRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.RenderMarkdownRequest.displayName = 'proto.gendocu.documentation.v1.RenderMarkdownRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.gendocu.documentation.v1.RenderMarkdownResponse.repeatedFields_, null);
};
goog.inherits(proto.gendocu.documentation.v1.RenderMarkdownResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.gendocu.documentation.v1.RenderMarkdownResponse.displayName = 'proto.gendocu.documentation.v1.RenderMarkdownResponse';
}



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    projectId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    previousSettingsVersion: jspb.Message.getFieldWithDefault(msg, 3, ""),
    settings: (f = msg.getSettings()) && proto.gendocu.documentation.v1.StylingSettings.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest;
  return proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setPreviousSettingsVersion(value);
      break;
    case 2:
      var value = new proto.gendocu.documentation.v1.StylingSettings;
      reader.readMessage(value,proto.gendocu.documentation.v1.StylingSettings.deserializeBinaryFromReader);
      msg.setSettings(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getPreviousSettingsVersion();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getSettings();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.gendocu.documentation.v1.StylingSettings.serializeBinaryToWriter
    );
  }
};


/**
 * optional string project_id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} returns this
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string previous_settings_version = 3;
 * @return {string}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.getPreviousSettingsVersion = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} returns this
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.setPreviousSettingsVersion = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional StylingSettings settings = 2;
 * @return {?proto.gendocu.documentation.v1.StylingSettings}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.getSettings = function() {
  return /** @type{?proto.gendocu.documentation.v1.StylingSettings} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.documentation.v1.StylingSettings, 2));
};


/**
 * @param {?proto.gendocu.documentation.v1.StylingSettings|undefined} value
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} returns this
*/
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.setSettings = function(value) {
  return jspb.Message.setWrapperField(this, 2, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest} returns this
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.clearSettings = function() {
  return this.setSettings(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingRequest.prototype.hasSettings = function() {
  return jspb.Message.getField(this, 2) != null;
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    projectId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    settings: (f = msg.getSettings()) && proto.gendocu.documentation.v1.StylingSettings.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse;
  return proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 2:
      var value = new proto.gendocu.documentation.v1.StylingSettings;
      reader.readMessage(value,proto.gendocu.documentation.v1.StylingSettings.deserializeBinaryFromReader);
      msg.setSettings(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getSettings();
  if (f != null) {
    writer.writeMessage(
      2,
      f,
      proto.gendocu.documentation.v1.StylingSettings.serializeBinaryToWriter
    );
  }
};


/**
 * optional string project_id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} returns this
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional StylingSettings settings = 2;
 * @return {?proto.gendocu.documentation.v1.StylingSettings}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.getSettings = function() {
  return /** @type{?proto.gendocu.documentation.v1.StylingSettings} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.documentation.v1.StylingSettings, 2));
};


/**
 * @param {?proto.gendocu.documentation.v1.StylingSettings|undefined} value
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} returns this
*/
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.setSettings = function(value) {
  return jspb.Message.setWrapperField(this, 2, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse} returns this
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.clearSettings = function() {
  return this.setSettings(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.documentation.v1.UpdateDocumentationStylingResponse.prototype.hasSettings = function() {
  return jspb.Message.getField(this, 2) != null;
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.documentation.v1.StylingSettings.repeatedFields_ = [5];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.StylingSettings.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.StylingSettings} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.StylingSettings.toObject = function(includeInstance, msg) {
  var f, obj = {
    settingsVersion: jspb.Message.getFieldWithDefault(msg, 3, ""),
    logoImg: jspb.Message.getFieldWithDefault(msg, 1, ""),
    logoWidth: jspb.Message.getFieldWithDefault(msg, 6, ""),
    logoHeight: jspb.Message.getFieldWithDefault(msg, 7, ""),
    documentationName: jspb.Message.getFieldWithDefault(msg, 2, ""),
    themeSettings: (f = msg.getThemeSettings()) && proto.gendocu.documentation.v1.ThemeSettings.toObject(includeInstance, f),
    customSectionsList: jspb.Message.toObjectList(msg.getCustomSectionsList(),
    gendocu_documentation_v1_sections_pb.CustomSection.toObject, includeInstance)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.StylingSettings}
 */
proto.gendocu.documentation.v1.StylingSettings.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.StylingSettings;
  return proto.gendocu.documentation.v1.StylingSettings.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.StylingSettings} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.StylingSettings}
 */
proto.gendocu.documentation.v1.StylingSettings.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 3:
      var value = /** @type {string} */ (reader.readString());
      msg.setSettingsVersion(value);
      break;
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setLogoImg(value);
      break;
    case 6:
      var value = /** @type {string} */ (reader.readString());
      msg.setLogoWidth(value);
      break;
    case 7:
      var value = /** @type {string} */ (reader.readString());
      msg.setLogoHeight(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setDocumentationName(value);
      break;
    case 4:
      var value = new proto.gendocu.documentation.v1.ThemeSettings;
      reader.readMessage(value,proto.gendocu.documentation.v1.ThemeSettings.deserializeBinaryFromReader);
      msg.setThemeSettings(value);
      break;
    case 5:
      var value = new gendocu_documentation_v1_sections_pb.CustomSection;
      reader.readMessage(value,gendocu_documentation_v1_sections_pb.CustomSection.deserializeBinaryFromReader);
      msg.addCustomSections(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.StylingSettings.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.StylingSettings} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.StylingSettings.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getSettingsVersion();
  if (f.length > 0) {
    writer.writeString(
      3,
      f
    );
  }
  f = message.getLogoImg();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getLogoWidth();
  if (f.length > 0) {
    writer.writeString(
      6,
      f
    );
  }
  f = message.getLogoHeight();
  if (f.length > 0) {
    writer.writeString(
      7,
      f
    );
  }
  f = message.getDocumentationName();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
  f = message.getThemeSettings();
  if (f != null) {
    writer.writeMessage(
      4,
      f,
      proto.gendocu.documentation.v1.ThemeSettings.serializeBinaryToWriter
    );
  }
  f = message.getCustomSectionsList();
  if (f.length > 0) {
    writer.writeRepeatedMessage(
      5,
      f,
      gendocu_documentation_v1_sections_pb.CustomSection.serializeBinaryToWriter
    );
  }
};


/**
 * optional string settings_version = 3;
 * @return {string}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getSettingsVersion = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.setSettingsVersion = function(value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};


/**
 * optional string logo_img = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getLogoImg = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.setLogoImg = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string logo_width = 6;
 * @return {string}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getLogoWidth = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.setLogoWidth = function(value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};


/**
 * optional string logo_height = 7;
 * @return {string}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getLogoHeight = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.setLogoHeight = function(value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};


/**
 * optional string documentation_name = 2;
 * @return {string}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getDocumentationName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.setDocumentationName = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};


/**
 * optional ThemeSettings theme_settings = 4;
 * @return {?proto.gendocu.documentation.v1.ThemeSettings}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getThemeSettings = function() {
  return /** @type{?proto.gendocu.documentation.v1.ThemeSettings} */ (
    jspb.Message.getWrapperField(this, proto.gendocu.documentation.v1.ThemeSettings, 4));
};


/**
 * @param {?proto.gendocu.documentation.v1.ThemeSettings|undefined} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
*/
proto.gendocu.documentation.v1.StylingSettings.prototype.setThemeSettings = function(value) {
  return jspb.Message.setWrapperField(this, 4, value);
};


/**
 * Clears the message field making it undefined.
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.clearThemeSettings = function() {
  return this.setThemeSettings(undefined);
};


/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.hasThemeSettings = function() {
  return jspb.Message.getField(this, 4) != null;
};


/**
 * repeated CustomSection custom_sections = 5;
 * @return {!Array<!proto.gendocu.documentation.v1.CustomSection>}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.getCustomSectionsList = function() {
  return /** @type{!Array<!proto.gendocu.documentation.v1.CustomSection>} */ (
    jspb.Message.getRepeatedWrapperField(this, gendocu_documentation_v1_sections_pb.CustomSection, 5));
};


/**
 * @param {!Array<!proto.gendocu.documentation.v1.CustomSection>} value
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
*/
proto.gendocu.documentation.v1.StylingSettings.prototype.setCustomSectionsList = function(value) {
  return jspb.Message.setRepeatedWrapperField(this, 5, value);
};


/**
 * @param {!proto.gendocu.documentation.v1.CustomSection=} opt_value
 * @param {number=} opt_index
 * @return {!proto.gendocu.documentation.v1.CustomSection}
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.addCustomSections = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 5, opt_value, proto.gendocu.documentation.v1.CustomSection, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.documentation.v1.StylingSettings} returns this
 */
proto.gendocu.documentation.v1.StylingSettings.prototype.clearCustomSectionsList = function() {
  return this.setCustomSectionsList([]);
};





if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.ThemeSettings.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.ThemeSettings} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.ThemeSettings.toObject = function(includeInstance, msg) {
  var f, obj = {
    themeId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    themeParams: jspb.Message.getFieldWithDefault(msg, 2, "")
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.ThemeSettings}
 */
proto.gendocu.documentation.v1.ThemeSettings.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.ThemeSettings;
  return proto.gendocu.documentation.v1.ThemeSettings.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.ThemeSettings} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.ThemeSettings}
 */
proto.gendocu.documentation.v1.ThemeSettings.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setThemeId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.setThemeParams(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.ThemeSettings.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.ThemeSettings} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.ThemeSettings.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getThemeId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getThemeParams();
  if (f.length > 0) {
    writer.writeString(
      2,
      f
    );
  }
};


/**
 * optional string theme_id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.getThemeId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.ThemeSettings} returns this
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.setThemeId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * optional string theme_params = 2;
 * @return {string}
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.getThemeParams = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.ThemeSettings} returns this
 */
proto.gendocu.documentation.v1.ThemeSettings.prototype.setThemeParams = function(value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.repeatedFields_ = [2];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.RenderMarkdownRequest.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownRequest} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.toObject = function(includeInstance, msg) {
  var f, obj = {
    projectId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    markdownList: (f = jspb.Message.getRepeatedField(msg, 2)) == null ? undefined : f
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.RenderMarkdownRequest;
  return proto.gendocu.documentation.v1.RenderMarkdownRequest.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.addMarkdown(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.RenderMarkdownRequest.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getMarkdownList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      2,
      f
    );
  }
};


/**
 * optional string project_id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * repeated string markdown = 2;
 * @return {!Array<string>}
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.getMarkdownList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 2));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.setMarkdownList = function(value) {
  return jspb.Message.setField(this, 2, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.addMarkdown = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 2, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownRequest} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownRequest.prototype.clearMarkdownList = function() {
  return this.setMarkdownList([]);
};



/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.repeatedFields_ = [2];



if (jspb.Message.GENERATE_TO_OBJECT) {
/**
 * Creates an object representation of this proto.
 * Field names that are reserved in JavaScript and will be renamed to pb_name.
 * Optional fields that are not set will be set to undefined.
 * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
 * For the list of reserved names please see:
 *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
 * @param {boolean=} opt_includeInstance Deprecated. whether to include the
 *     JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @return {!Object}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.toObject = function(opt_includeInstance) {
  return proto.gendocu.documentation.v1.RenderMarkdownResponse.toObject(opt_includeInstance, this);
};


/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownResponse} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.toObject = function(includeInstance, msg) {
  var f, obj = {
    projectId: jspb.Message.getFieldWithDefault(msg, 1, ""),
    htmlList: (f = jspb.Message.getRepeatedField(msg, 2)) == null ? undefined : f
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}


/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.gendocu.documentation.v1.RenderMarkdownResponse;
  return proto.gendocu.documentation.v1.RenderMarkdownResponse.deserializeBinaryFromReader(msg, reader);
};


/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {string} */ (reader.readString());
      msg.setProjectId(value);
      break;
    case 2:
      var value = /** @type {string} */ (reader.readString());
      msg.addHtml(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};


/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter();
  proto.gendocu.documentation.v1.RenderMarkdownResponse.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};


/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.gendocu.documentation.v1.RenderMarkdownResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.serializeBinaryToWriter = function(message, writer) {
  var f = undefined;
  f = message.getProjectId();
  if (f.length > 0) {
    writer.writeString(
      1,
      f
    );
  }
  f = message.getHtmlList();
  if (f.length > 0) {
    writer.writeRepeatedString(
      2,
      f
    );
  }
};


/**
 * optional string project_id = 1;
 * @return {string}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.getProjectId = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};


/**
 * @param {string} value
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.setProjectId = function(value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};


/**
 * repeated string html = 2;
 * @return {!Array<string>}
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.getHtmlList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 2));
};


/**
 * @param {!Array<string>} value
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.setHtmlList = function(value) {
  return jspb.Message.setField(this, 2, value || []);
};


/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.addHtml = function(value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 2, value, opt_index);
};


/**
 * Clears the list making it empty but non-null.
 * @return {!proto.gendocu.documentation.v1.RenderMarkdownResponse} returns this
 */
proto.gendocu.documentation.v1.RenderMarkdownResponse.prototype.clearHtmlList = function() {
  return this.setHtmlList([]);
};


goog.object.extend(exports, proto.gendocu.documentation.v1);
