// @useoptic/optic/build/client/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnauthorizedError = exports.ServiceUnavailableError = exports.NotFoundError = exports.InternalError = exports.ForbiddenError = exports.BadRequestError = void 0;
class BadRequestError extends Error {
  constructor(msg, source) {
    super(msg);
    this.code = 400;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, BadRequestError.prototype);
  }
}
exports.BadRequestError = BadRequestError;
class UnauthorizedError extends Error {
  constructor(msg, source) {
    super(msg);
    this.code = 401;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, UnauthorizedError.prototype);
  }
}
exports.UnauthorizedError = UnauthorizedError;
class ForbiddenError extends Error {
  constructor(msg, source) {
    super(msg);
    this.code = 403;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, ForbiddenError.prototype);
  }
}
exports.ForbiddenError = ForbiddenError;
class NotFoundError extends Error {
  constructor(msg, source = 'Not found') {
    super(msg);
    this.code = 404;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, NotFoundError.prototype);
  }
}
exports.NotFoundError = NotFoundError;
class InternalError extends Error {
  constructor(msg, source) {
    super(msg);
    this.code = 500;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, InternalError.prototype);
  }
}
exports.InternalError = InternalError;
class ServiceUnavailableError extends Error {
  constructor(msg, source) {
    super(msg);
    this.code = 503;
    this.source = source;
    // https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, ServiceUnavailableError.prototype);
  }
}
exports.ServiceUnavailableError = ServiceUnavailableError;
