// @useoptic/rulesets-base/build/test-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptySpec = exports.externalRulesWithInputs = exports.runRulesWithInputs = exports.createRuleInputs = void 0;
const openapi_utilities_1 = require("@useoptic/openapi-utilities");
const rule_runner_1 = require("./rule-runner");
const createRuleInputs = (beforeJson, afterJson) => {
    const currentTraverser = new openapi_utilities_1.OpenAPITraverser();
    const nextTraverser = new openapi_utilities_1.OpenAPITraverser();
    currentTraverser.traverse(beforeJson);
    const currentFacts = [...currentTraverser.facts()];
    nextTraverser.traverse(afterJson);
    const nextFacts = [...nextTraverser.facts()];
    const changes = (0, openapi_utilities_1.factsToChangelog)(currentFacts, nextFacts);
    return {
        context: {},
        currentFacts,
        nextFacts,
        changelog: changes,
        currentJsonLike: beforeJson,
        nextJsonLike: afterJson,
    };
};
exports.createRuleInputs = createRuleInputs;
const runRulesWithInputs = (rules, beforeJson, afterJson) => {
    const ruleRunner = new rule_runner_1.RuleRunner(rules);
    return ruleRunner.runRulesWithFacts((0, exports.createRuleInputs)(beforeJson, afterJson));
};
exports.runRulesWithInputs = runRulesWithInputs;
const externalRulesWithInputs = (rules, beforeJson, afterJson) => {
    const ruleRunner = new rule_runner_1.RuleRunner([rules]);
    return ruleRunner.runRulesWithFacts((0, exports.createRuleInputs)(beforeJson, afterJson));
};
exports.externalRulesWithInputs = externalRulesWithInputs;
const createEmptySpec = () => ({
    ...openapi_utilities_1.defaultEmptySpec,
    paths: {},
    info: {
        ...openapi_utilities_1.defaultEmptySpec.info,
    },
});
exports.createEmptySpec = createEmptySpec;
